{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyreadr\n",
    "import polars as pl\n",
    "import polars.selectors as cs\n",
    "\n",
    "diffexp_x = pl.read_csv(snakemake.input[0], separator=\"\\t\").lazy()\n",
    "diffexp_y = pl.read_csv(snakemake.input[1], separator=\"\\t\").lazy()\n",
    "label_x = snakemake.params.labels[0]\n",
    "label_y = snakemake.params.labels[1]\n",
    "\n",
    "# diffexp_x = pl.read_csv(\"/projects/koesterlab/moeller-th-liver-diffexp/moeller-rna-liver-2024/results/tables/pathways/etoh_mgl_in_wt_vs_etho_in_wt_liver.pathways.tsv\", separator=\"\\t\").lazy()\n",
    "# diffexp_y = pl.read_csv(\"/projects/koesterlab/moeller-th-liver-diffexp/moeller-rna-liver-2024/results/tables/pathways/etoh_t3_in_wt_vs_etoh_in_wt_liver.pathways.tsv\", separator=\"\\t\").lazy()\n",
    "# label_x = \"mgl\"\n",
    "# label_y = \"t3\"\n",
    "\n",
    "effect_x = f\"effect {label_x}\"\n",
    "effect_y = f\"effect {label_y}\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare(df):\n",
    "    # Select necessary columns and filter\n",
    "    df = df.select([\n",
    "               cs.by_name(\"Name\", \"Combined p-value\", \"Combined FDR\", \"total perturbation accumulation\", \"pathway id\")\n",
    "           ])\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "prepared_diffexp_x = prepare(diffexp_x)\n",
    "prepared_diffexp_y = prepare(diffexp_y)\n",
    "combined = prepared_diffexp_x.join(\n",
    "    prepared_diffexp_y, on=[\"Name\"], suffix=\"_y\"\n",
    ").with_columns(\n",
    "    pl.min_horizontal(\"Combined FDR\", \"Combined FDR_y\").alias(\"fdr_min\")\n",
    ").filter(\n",
    "    pl.col(\"fdr_min\") <= 0.05\n",
    ").rename(\n",
    "    {\n",
    "        \"total perturbation accumulation\": effect_x,\n",
    "        \"total perturbation accumulation_y\": effect_y,\n",
    "        \"fdr_min\": \"min fdr\",\n",
    "    }\n",
    ").collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (426, 10)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Name</th><th>Combined p-value</th><th>Combined FDR</th><th>effect mgl</th><th>pathway id</th><th>Combined p-value_y</th><th>Combined FDR_y</th><th>effect t3</th><th>pathway id_y</th><th>min fdr</th></tr><tr><td>str</td><td>f64</td><td>f64</td><td>f64</td><td>str</td><td>f64</td><td>f64</td><td>f64</td><td>str</td><td>f64</td></tr></thead><tbody><tr><td>&quot;Metabolism of RNA&quot;</td><td>3.6577e-104</td><td>2.7835e-101</td><td>-153.047343</td><td>&quot;R-MMU-8953854&quot;</td><td>0.000008</td><td>0.000236</td><td>-47.991535</td><td>&quot;R-MMU-8953854&quot;</td><td>2.7835e-101</td></tr><tr><td>&quot;Translation&quot;</td><td>3.7171e-69</td><td>1.4143e-66</td><td>-25.400783</td><td>&quot;R-MMU-72766&quot;</td><td>0.025606</td><td>0.102425</td><td>-7.763833</td><td>&quot;R-MMU-72766&quot;</td><td>1.4143e-66</td></tr><tr><td>&quot;Amino acid and derivative meta…</td><td>3.1699e-51</td><td>7.4966e-49</td><td>8.696646</td><td>&quot;R-MMU-71291&quot;</td><td>1.1662e-24</td><td>3.5452e-22</td><td>-1.63</td><td>&quot;R-MMU-71291&quot;</td><td>7.4966e-49</td></tr><tr><td>&quot;rRNA processing&quot;</td><td>3.9404e-51</td><td>7.4966e-49</td><td>-140.95763</td><td>&quot;R-MMU-72312&quot;</td><td>0.000175</td><td>0.003048</td><td>-37.557806</td><td>&quot;R-MMU-72312&quot;</td><td>7.4966e-49</td></tr><tr><td>&quot;rRNA processing in the nucleus…</td><td>5.8909e-51</td><td>8.9659e-49</td><td>-139.377208</td><td>&quot;R-MMU-8868773&quot;</td><td>0.000188</td><td>0.003176</td><td>-37.46126</td><td>&quot;R-MMU-8868773&quot;</td><td>8.9659e-49</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;FCERI mediated Ca+2 mobilizati…</td><td>0.21634</td><td>0.27763</td><td>1.655</td><td>&quot;R-MMU-2871809&quot;</td><td>0.002636</td><td>0.026708</td><td>7.439</td><td>&quot;R-MMU-2871809&quot;</td><td>0.026708</td></tr><tr><td>&quot;DAP12 interactions&quot;</td><td>0.227572</td><td>0.288637</td><td>3.038273</td><td>&quot;R-MMU-2172127&quot;</td><td>0.005966</td><td>0.046887</td><td>5.972909</td><td>&quot;R-MMU-2172127&quot;</td><td>0.046887</td></tr><tr><td>&quot;Interleukin-3, Interleukin-5 a…</td><td>0.344125</td><td>0.40087</td><td>0.0</td><td>&quot;R-MMU-512988&quot;</td><td>0.003173</td><td>0.02905</td><td>1.02</td><td>&quot;R-MMU-512988&quot;</td><td>0.02905</td></tr><tr><td>&quot;Collagen formation&quot;</td><td>0.419314</td><td>0.476266</td><td>1.26257</td><td>&quot;R-MMU-1474290&quot;</td><td>0.004462</td><td>0.03743</td><td>-15.229161</td><td>&quot;R-MMU-1474290&quot;</td><td>0.03743</td></tr><tr><td>&quot;G alpha (q) signalling events&quot;</td><td>0.685333</td><td>0.73147</td><td>3.21955</td><td>&quot;R-MMU-416476&quot;</td><td>0.006453</td><td>0.049045</td><td>9.114</td><td>&quot;R-MMU-416476&quot;</td><td>0.049045</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (426, 10)\n",
       "┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐\n",
       "│ Name      ┆ Combined  ┆ Combined  ┆ effect    ┆ … ┆ Combined  ┆ effect t3 ┆ pathway   ┆ min fdr  │\n",
       "│ ---       ┆ p-value   ┆ FDR       ┆ mgl       ┆   ┆ FDR_y     ┆ ---       ┆ id_y      ┆ ---      │\n",
       "│ str       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ f64       ┆ ---       ┆ f64      │\n",
       "│           ┆ f64       ┆ f64       ┆ f64       ┆   ┆ f64       ┆           ┆ str       ┆          │\n",
       "╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡\n",
       "│ Metabolis ┆ 3.6577e-1 ┆ 2.7835e-1 ┆ -153.0473 ┆ … ┆ 0.000236  ┆ -47.99153 ┆ R-MMU-895 ┆ 2.7835e- │\n",
       "│ m of RNA  ┆ 04        ┆ 01        ┆ 43        ┆   ┆           ┆ 5         ┆ 3854      ┆ 101      │\n",
       "│ Translati ┆ 3.7171e-6 ┆ 1.4143e-6 ┆ -25.40078 ┆ … ┆ 0.102425  ┆ -7.763833 ┆ R-MMU-727 ┆ 1.4143e- │\n",
       "│ on        ┆ 9         ┆ 6         ┆ 3         ┆   ┆           ┆           ┆ 66        ┆ 66       │\n",
       "│ Amino     ┆ 3.1699e-5 ┆ 7.4966e-4 ┆ 8.696646  ┆ … ┆ 3.5452e-2 ┆ -1.63     ┆ R-MMU-712 ┆ 7.4966e- │\n",
       "│ acid and  ┆ 1         ┆ 9         ┆           ┆   ┆ 2         ┆           ┆ 91        ┆ 49       │\n",
       "│ derivativ ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ e meta…   ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ rRNA proc ┆ 3.9404e-5 ┆ 7.4966e-4 ┆ -140.9576 ┆ … ┆ 0.003048  ┆ -37.55780 ┆ R-MMU-723 ┆ 7.4966e- │\n",
       "│ essing    ┆ 1         ┆ 9         ┆ 3         ┆   ┆           ┆ 6         ┆ 12        ┆ 49       │\n",
       "│ rRNA proc ┆ 5.8909e-5 ┆ 8.9659e-4 ┆ -139.3772 ┆ … ┆ 0.003176  ┆ -37.46126 ┆ R-MMU-886 ┆ 8.9659e- │\n",
       "│ essing in ┆ 1         ┆ 9         ┆ 08        ┆   ┆           ┆           ┆ 8773      ┆ 49       │\n",
       "│ the       ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ nucleus…  ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ …         ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …         ┆ …        │\n",
       "│ FCERI     ┆ 0.21634   ┆ 0.27763   ┆ 1.655     ┆ … ┆ 0.026708  ┆ 7.439     ┆ R-MMU-287 ┆ 0.026708 │\n",
       "│ mediated  ┆           ┆           ┆           ┆   ┆           ┆           ┆ 1809      ┆          │\n",
       "│ Ca+2 mobi ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ lizati…   ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ DAP12 int ┆ 0.227572  ┆ 0.288637  ┆ 3.038273  ┆ … ┆ 0.046887  ┆ 5.972909  ┆ R-MMU-217 ┆ 0.046887 │\n",
       "│ eractions ┆           ┆           ┆           ┆   ┆           ┆           ┆ 2127      ┆          │\n",
       "│ Interleuk ┆ 0.344125  ┆ 0.40087   ┆ 0.0       ┆ … ┆ 0.02905   ┆ 1.02      ┆ R-MMU-512 ┆ 0.02905  │\n",
       "│ in-3, Int ┆           ┆           ┆           ┆   ┆           ┆           ┆ 988       ┆          │\n",
       "│ erleukin- ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ 5 a…      ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ Collagen  ┆ 0.419314  ┆ 0.476266  ┆ 1.26257   ┆ … ┆ 0.03743   ┆ -15.22916 ┆ R-MMU-147 ┆ 0.03743  │\n",
       "│ formation ┆           ┆           ┆           ┆   ┆           ┆ 1         ┆ 4290      ┆          │\n",
       "│ G alpha   ┆ 0.685333  ┆ 0.73147   ┆ 3.21955   ┆ … ┆ 0.049045  ┆ 9.114     ┆ R-MMU-416 ┆ 0.049045 │\n",
       "│ (q) signa ┆           ┆           ┆           ┆   ┆           ┆           ┆ 476       ┆          │\n",
       "│ lling     ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ events    ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "pa.Table requires 'pyarrow' module to be installed",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 10\u001b[0m\n\u001b[1;32m      4\u001b[0m combined \u001b[38;5;241m=\u001b[39m combined\u001b[38;5;241m.\u001b[39mwith_columns(\n\u001b[1;32m      5\u001b[0m     \u001b[38;5;28mabs\u001b[39m(pl\u001b[38;5;241m.\u001b[39mcol(effect_x) \u001b[38;5;241m-\u001b[39m pl\u001b[38;5;241m.\u001b[39mcol(effect_y))\u001b[38;5;241m.\u001b[39malias(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdifference\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      6\u001b[0m )\n\u001b[1;32m      7\u001b[0m combined_sorted \u001b[38;5;241m=\u001b[39m combined\u001b[38;5;241m.\u001b[39msort(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdifference\u001b[39m\u001b[38;5;124m\"\u001b[39m, descending\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m      8\u001b[0m combined_pd \u001b[38;5;241m=\u001b[39m \u001b[43mcombined_sorted\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mselect\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      9\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpl\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcol\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mName\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmin fdr\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43meffect_x\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43meffect_y\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mdifference\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m---> 10\u001b[0m \u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_pandas\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     11\u001b[0m combined_pd\u001b[38;5;241m.\u001b[39mto_csv(snakemake\u001b[38;5;241m.\u001b[39moutput[\u001b[38;5;241m0\u001b[39m], sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m\"\u001b[39m, index\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m     12\u001b[0m \u001b[38;5;66;03m# combined_pd.to_csv(\"/projects/koesterlab/moeller-th-liver-diffexp/analysis_2/results/tables/pathways/meta_compare_etoh_wt_mgl_vs_t3.tsv\", sep=\"\\t\", index=False)\u001b[39;00m\n",
      "File \u001b[0;32m/projects/envs/conda/aprinz/envs/jupyter/lib/python3.12/site-packages/polars/dataframe/frame.py:2023\u001b[0m, in \u001b[0;36mDataFrame.to_pandas\u001b[0;34m(self, use_pyarrow_extension_array, **kwargs)\u001b[0m\n\u001b[1;32m   2018\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m Object \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdtypes:\n\u001b[1;32m   2019\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_to_pandas_with_object_columns(\n\u001b[1;32m   2020\u001b[0m         use_pyarrow_extension_array\u001b[38;5;241m=\u001b[39muse_pyarrow_extension_array, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs\n\u001b[1;32m   2021\u001b[0m     )\n\u001b[0;32m-> 2023\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_to_pandas_without_object_columns\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   2024\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43muse_pyarrow_extension_array\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43muse_pyarrow_extension_array\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\n\u001b[1;32m   2025\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/projects/envs/conda/aprinz/envs/jupyter/lib/python3.12/site-packages/polars/dataframe/frame.py:2075\u001b[0m, in \u001b[0;36mDataFrame._to_pandas_without_object_columns\u001b[0;34m(self, df, use_pyarrow_extension_array, **kwargs)\u001b[0m\n\u001b[1;32m   2072\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m pd\u001b[38;5;241m.\u001b[39mDataFrame()\n\u001b[1;32m   2074\u001b[0m record_batches \u001b[38;5;241m=\u001b[39m df\u001b[38;5;241m.\u001b[39m_df\u001b[38;5;241m.\u001b[39mto_pandas()\n\u001b[0;32m-> 2075\u001b[0m tbl \u001b[38;5;241m=\u001b[39m \u001b[43mpa\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mTable\u001b[49m\u001b[38;5;241m.\u001b[39mfrom_batches(record_batches)\n\u001b[1;32m   2076\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m use_pyarrow_extension_array:\n\u001b[1;32m   2077\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m tbl\u001b[38;5;241m.\u001b[39mto_pandas(\n\u001b[1;32m   2078\u001b[0m         self_destruct\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m   2079\u001b[0m         split_blocks\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m   2080\u001b[0m         types_mapper\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mlambda\u001b[39;00m pa_dtype: pd\u001b[38;5;241m.\u001b[39mArrowDtype(pa_dtype),\n\u001b[1;32m   2081\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m   2082\u001b[0m     )\n",
      "File \u001b[0;32m/projects/envs/conda/aprinz/envs/jupyter/lib/python3.12/site-packages/polars/dependencies.py:97\u001b[0m, in \u001b[0;36m_LazyModule.__getattr__\u001b[0;34m(self, attr)\u001b[0m\n\u001b[1;32m     95\u001b[0m pfx \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_mod_pfx\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_module_name, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     96\u001b[0m msg \u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpfx\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;132;01m{\u001b[39;00mattr\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m requires \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_module_name\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m module to be installed\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m---> 97\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mModuleNotFoundError\u001b[39;00m(msg) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: pa.Table requires 'pyarrow' module to be installed"
     ]
    }
   ],
   "source": [
    "effects = combined.select(pl.col(effect_x, effect_y))\n",
    "min_value = effects.min().min_horizontal()[0]\n",
    "max_value = effects.max().max_horizontal()[0]\n",
    "combined = combined.with_columns(\n",
    "    abs(pl.col(effect_x) - pl.col(effect_y)).alias(\"difference\")\n",
    ")\n",
    "combined_sorted = combined.sort(\"difference\", descending=True)\n",
    "combined_pd = combined_sorted.select(\n",
    "    pl.col(\"Name\", \"min fdr\", effect_x, effect_y, \"difference\", \"pathway id\")\n",
    ").to_pandas()\n",
    "combined_pd.to_csv(snakemake.output[0], sep=\"\\t\", index=False)\n",
    "# combined_pd.to_csv(\"/projects/koesterlab/moeller-th-liver-diffexp/analysis_2/results/tables/pathways/meta_compare_etoh_wt_mgl_vs_t3.tsv\", sep=\"\\t\", index=False)\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<style>\n",
       "  #altair-viz-53b609c2906f4e289f9eecf06c667b9d.vega-embed {\n",
       "    width: 100%;\n",
       "    display: flex;\n",
       "  }\n",
       "\n",
       "  #altair-viz-53b609c2906f4e289f9eecf06c667b9d.vega-embed details,\n",
       "  #altair-viz-53b609c2906f4e289f9eecf06c667b9d.vega-embed details summary {\n",
       "    position: relative;\n",
       "  }\n",
       "</style>\n",
       "<div id=\"altair-viz-53b609c2906f4e289f9eecf06c667b9d\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-53b609c2906f4e289f9eecf06c667b9d\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-53b609c2906f4e289f9eecf06c667b9d\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm/vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm/vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm/vega-lite@5.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm/vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"5.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 300, \"continuousHeight\": 300}}, \"layer\": [{\"data\": {\"name\": \"data-5604c526602009a8abb16b022d3672a3\"}, \"mark\": {\"type\": \"line\", \"color\": \"lightgrey\"}, \"encoding\": {\"strokeDash\": {\"value\": [5, 5]}, \"x\": {\"field\": \"effect mgl\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"effect t3\", \"type\": \"quantitative\"}}, \"name\": \"view_5\"}, {\"data\": {\"name\": \"data-938864a5c849e86e9fc1a3da27ef7105\"}, \"mark\": {\"type\": \"line\", \"color\": \"lightgrey\"}, \"encoding\": {\"strokeDash\": {\"value\": [5, 5]}, \"x\": {\"field\": \"effect mgl\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"effect t3\", \"type\": \"quantitative\"}}}, {\"data\": {\"name\": \"data-c1e4579224ac97fc05b1e721fa5c8ed9\"}, \"mark\": {\"type\": \"line\", \"color\": \"lightgrey\"}, \"encoding\": {\"strokeDash\": {\"value\": [5, 5]}, \"x\": {\"field\": \"effect mgl\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"effect t3\", \"type\": \"quantitative\"}}}, {\"data\": {\"name\": \"data-9f19acd278933992bd977d08806fce4b\"}, \"mark\": {\"type\": \"circle\", \"size\": 15, \"tooltip\": {\"content\": \"data\"}}, \"encoding\": {\"color\": {\"field\": \"min fdr\", \"scale\": {\"scheme\": \"viridis\"}, \"type\": \"quantitative\"}, \"opacity\": {\"value\": 0.5}, \"x\": {\"axis\": {\"grid\": false}, \"field\": \"effect mgl\", \"scale\": {\"nice\": false, \"type\": \"symlog\"}, \"type\": \"quantitative\"}, \"y\": {\"axis\": {\"grid\": false}, \"field\": \"effect t3\", \"scale\": {\"nice\": false, \"type\": \"symlog\"}, \"type\": \"quantitative\"}}}, {\"data\": {\"name\": \"data-9f19acd278933992bd977d08806fce4b\"}, \"mark\": {\"type\": \"text\", \"align\": \"left\", \"baseline\": \"middle\", \"dx\": 5, \"dy\": -5, \"fill\": \"white\", \"stroke\": \"white\", \"strokeWidth\": 5}, \"encoding\": {\"text\": {\"condition\": {\"param\": \"param_9\", \"empty\": false, \"field\": \"Name\"}, \"value\": \"\"}, \"x\": {\"field\": \"effect mgl\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"effect t3\", \"type\": \"quantitative\"}}}, {\"data\": {\"name\": \"data-9f19acd278933992bd977d08806fce4b\"}, \"mark\": {\"type\": \"text\", \"align\": \"left\", \"baseline\": \"middle\", \"dx\": 5, \"dy\": -5}, \"encoding\": {\"text\": {\"condition\": {\"param\": \"param_9\", \"empty\": false, \"field\": \"Name\"}, \"value\": \"\"}, \"x\": {\"field\": \"effect mgl\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"effect t3\", \"type\": \"quantitative\"}}}], \"params\": [{\"name\": \"param_9\", \"select\": {\"type\": \"point\", \"fields\": [\"Name\"]}, \"views\": [\"view_5\"]}, {\"name\": \"param_10\", \"select\": {\"type\": \"interval\", \"encodings\": [\"x\", \"y\"]}, \"bind\": \"scales\", \"views\": [\"view_5\"]}], \"$schema\": \"https://vega.github.io/schema/vega-lite/v5.17.0.json\", \"datasets\": {\"data-5604c526602009a8abb16b022d3672a3\": [{\"effect mgl\": 0, \"effect t3\": -283.1369967850618}, {\"effect mgl\": 0, \"effect t3\": 103216.48991414806}], \"data-938864a5c849e86e9fc1a3da27ef7105\": [{\"effect mgl\": -283.1369967850618, \"effect t3\": 0}, {\"effect mgl\": 103216.48991414806, \"effect t3\": 0}], \"data-c1e4579224ac97fc05b1e721fa5c8ed9\": [{\"effect mgl\": -283.1369967850618, \"effect t3\": -283.1369967850618}, {\"effect mgl\": 103216.48991414806, \"effect t3\": 103216.48991414806}], \"data-9f19acd278933992bd977d08806fce4b\": [{\"Name\": \"Mitotic Metaphase and Anaphase\", \"min fdr\": 6.46116843533494e-21, \"effect mgl\": 103057.99505707111, \"effect t3\": 69072.93776539483}, {\"Name\": \"Mitotic Anaphase\", \"min fdr\": 4.752720799178139e-21, \"effect mgl\": 103216.48991414806, \"effect t3\": 69583.52954086108}, {\"Name\": \"Separation of Sister Chromatids\", \"min fdr\": 4.399211051290806e-21, \"effect mgl\": 102139.51308043645, \"effect t3\": 69423.97657144774}, {\"Name\": \"APC:Cdc20 mediated degradation of cell cycle proteins prior to satisfation of the cell cycle checkpoint\", \"min fdr\": 9.700366089373443e-21, \"effect mgl\": 1438.506949569788, \"effect t3\": 681.5271470588187}, {\"Name\": \"Developmental Biology\", \"min fdr\": 1.3380442297540586e-06, \"effect mgl\": 559.5023781775853, \"effect t3\": -1.0373449695667798}, {\"Name\": \"APC/C:Cdc20 mediated degradation of mitotic proteins\", \"min fdr\": 4.5311070567821995e-20, \"effect mgl\": 804.6651119480424, \"effect t3\": 406.8329641763517}, {\"Name\": \"Intra-Golgi and retrograde Golgi-to-ER traffic\", \"min fdr\": 3.5270042407376597e-14, \"effect mgl\": 345.2981499999949, \"effect t3\": 2.5779999999999994}, {\"Name\": \"Activation of APC/C and APC/C:Cdc20 mediated degradation of mitotic proteins\", \"min fdr\": 8.762146222976448e-20, \"effect mgl\": 630.3990848927868, \"effect t3\": 324.20638998069023}, {\"Name\": \"Antigen processing: Ubiquitination & Proteasome degradation\", \"min fdr\": 5.781317458740795e-37, \"effect mgl\": -279.9127601116376, \"effect t3\": -8.146245603015188}, {\"Name\": \"Class I MHC mediated antigen processing & presentation\", \"min fdr\": 1.0730777549302047e-40, \"effect mgl\": -283.1369967850618, \"effect t3\": -36.185113864778444}, {\"Name\": \"Downregulation of ERBB2 signaling\", \"min fdr\": 0.0017728771759552631, \"effect mgl\": 123.7487142857142, \"effect t3\": -69.85321428571423}, {\"Name\": \"Beta-catenin independent WNT signaling\", \"min fdr\": 1.2564564656914095e-15, \"effect mgl\": 168.57760208333318, \"effect t3\": 7.082500000000001}, {\"Name\": \"PCP/CE pathway\", \"min fdr\": 1.598066099685261e-17, \"effect mgl\": 164.9812749999998, \"effect t3\": 6.212500000000001}, {\"Name\": \"Hedgehog ligand biogenesis\", \"min fdr\": 9.471133350447654e-22, \"effect mgl\": 159.65047248226972, \"effect t3\": 7.228134751773062}, {\"Name\": \"CLEC7A (Dectin-1) signaling\", \"min fdr\": 3.593769354077238e-22, \"effect mgl\": 209.03353381038934, \"effect t3\": 60.72178779220772}, {\"Name\": \"C-type lectin receptors (CLRs)\", \"min fdr\": 3.952427643334665e-22, \"effect mgl\": 208.09418112207743, \"effect t3\": 64.57460839826825}, {\"Name\": \"Signaling by Rho GTPases, Miro GTPases and RHOBTB3\", \"min fdr\": 2.7227489681280476e-26, \"effect mgl\": 178.46923854614798, \"effect t3\": 40.21981988778582}, {\"Name\": \"Signaling by Rho GTPases\", \"min fdr\": 6.284316211935562e-26, \"effect mgl\": 175.5394014667297, \"effect t3\": 48.79968085924283}, {\"Name\": \"Plasma lipoprotein clearance\", \"min fdr\": 2.480613011220674e-11, \"effect mgl\": 122.18463714285708, \"effect t3\": 0.0}, {\"Name\": \"Signaling by WNT\", \"min fdr\": 2.807181279300746e-17, \"effect mgl\": 123.57599156033645, \"effect t3\": 6.699592540096212}, {\"Name\": \"rRNA processing\", \"min fdr\": 1.0352356394617035e-48, \"effect mgl\": -140.04937348362532, \"effect t3\": -38.00545201313544}, {\"Name\": \"Major pathway of rRNA processing in the nucleolus and cytosol\", \"min fdr\": 9.339248936449581e-49, \"effect mgl\": -138.44223059728245, \"effect t3\": -36.96517013587135}, {\"Name\": \"rRNA processing in the nucleus and cytosol\", \"min fdr\": 2.730914879801088e-51, \"effect mgl\": -138.86294069984214, \"effect t3\": -37.756991849685804}, {\"Name\": \"Metabolism of vitamins and cofactors\", \"min fdr\": 1.8056855388170625e-24, \"effect mgl\": 122.23181250000002, \"effect t3\": 21.842416666666676}, {\"Name\": \"Metabolism of RNA\", \"min fdr\": 3.8804243224765774e-101, \"effect mgl\": -147.8280483988789, \"effect t3\": -47.932192301541775}, {\"Name\": \"Nuclear events mediated by NFE2L2\", \"min fdr\": 3.5452221837087663e-22, \"effect mgl\": 137.25148819727949, \"effect t3\": 46.77072643343065}, {\"Name\": \"Signaling by Hedgehog\", \"min fdr\": 1.2564564656914095e-15, \"effect mgl\": 101.81274639775874, \"effect t3\": 13.17215789807765}, {\"Name\": \"GSK3B and BTRC:CUL1-mediated-degradation of NFE2L2\", \"min fdr\": 1.1090445489322008e-22, \"effect mgl\": 134.4399782604475, \"effect t3\": 48.229274295432575}, {\"Name\": \"MHC class II antigen presentation\", \"min fdr\": 4.559748087704435e-11, \"effect mgl\": 96.11049999999992, \"effect t3\": 14.944999999999986}, {\"Name\": \"Innate Immune System\", \"min fdr\": 2.781311744566336e-35, \"effect mgl\": 111.69885472806047, \"effect t3\": 39.19253830321981}, {\"Name\": \"Macroautophagy\", \"min fdr\": 1.4511086525125242e-15, \"effect mgl\": -63.806086788211786, \"effect t3\": 7.522630702630709}, {\"Name\": \"Autophagy\", \"min fdr\": 1.4277080800523268e-15, \"effect mgl\": -62.365273476523484, \"effect t3\": 8.008278346653356}, {\"Name\": \"Non-coding RNA Metabolism\", \"min fdr\": 0.0008501593540296957, \"effect mgl\": -84.58805333333339, \"effect t3\": -17.092000000000024}, {\"Name\": \"snRNP Assembly\", \"min fdr\": 0.0008751665259245636, \"effect mgl\": -80.85861500000006, \"effect t3\": -15.629833333333366}, {\"Name\": \"Transport of Mature mRNA derived from an Intron-Containing Transcript\", \"min fdr\": 6.677109506949697e-10, \"effect mgl\": 41.02420593103442, \"effect t3\": -23.88636445623341}, {\"Name\": \"Orc1 removal from chromatin\", \"min fdr\": 1.275410857379683e-19, \"effect mgl\": 109.84548829629632, \"effect t3\": 45.44316296296297}, {\"Name\": \"DNA Replication\", \"min fdr\": 4.681514001992429e-20, \"effect mgl\": 190.368476728769, \"effect t3\": 126.71634332637285}, {\"Name\": \"LDL clearance\", \"min fdr\": 2.915898317674781e-06, \"effect mgl\": 61.554500000000004, \"effect t3\": 0.0}, {\"Name\": \"GLI3 is processed to GLI3R by the proteasome\", \"min fdr\": 4.300731887782419e-22, \"effect mgl\": 92.7660779999996, \"effect t3\": 31.340849999999868}, {\"Name\": \"ESR-mediated signaling\", \"min fdr\": 0.007228810315513243, \"effect mgl\": 29.727546874999934, \"effect t3\": 82.7863125}, {\"Name\": \"Retinoid metabolism and transport\", \"min fdr\": 1.3880088911270077e-07, \"effect mgl\": 77.90912500000005, \"effect t3\": 25.03300000000002}, {\"Name\": \"Metabolism of fat-soluble vitamins\", \"min fdr\": 1.3054871801389165e-07, \"effect mgl\": 77.67017500000004, \"effect t3\": 25.48325000000002}, {\"Name\": \"Visual phototransduction\", \"min fdr\": 9.077963411486252e-08, \"effect mgl\": 76.09265000000005, \"effect t3\": 26.77375000000002}, {\"Name\": \"Signaling by Receptor Tyrosine Kinases\", \"min fdr\": 3.0473767677621927e-12, \"effect mgl\": 38.66926894361651, \"effect t3\": -9.79821523269774}, {\"Name\": \"Mitotic G2-G2/M phases\", \"min fdr\": 1.9228360505429123e-21, \"effect mgl\": 73.42065449308525, \"effect t3\": 26.47155105485533}, {\"Name\": \"Signaling by Nuclear Receptors\", \"min fdr\": 2.251454273648813e-06, \"effect mgl\": 23.786845238095268, \"effect t3\": 69.07782142857138}, {\"Name\": \"G2/M Transition\", \"min fdr\": 1.8635886182736746e-20, \"effect mgl\": 71.62929145979662, \"effect t3\": 27.200152779656655}, {\"Name\": \"Toll Like Receptor 3 (TLR3) Cascade\", \"min fdr\": 4.301630095275583e-08, \"effect mgl\": -75.01343145944175, \"effect t3\": -31.984726149970292}, {\"Name\": \"MyD88 cascade initiated on plasma membrane\", \"min fdr\": 4.6706690998189513e-08, \"effect mgl\": -72.91091550681593, \"effect t3\": -30.71667911587665}, {\"Name\": \"Toll Like Receptor 2 (TLR2) Cascade\", \"min fdr\": 4.1779064327681274e-08, \"effect mgl\": -72.97602250101978, \"effect t3\": -31.652607446105392}, {\"Name\": \"Toll Like Receptor TLR6:TLR2 Cascade\", \"min fdr\": 4.106765903698828e-08, \"effect mgl\": -72.42966547886654, \"effect t3\": -31.207569571994938}, {\"Name\": \"MyD88:MAL(TIRAP) cascade initiated on plasma membrane\", \"min fdr\": 3.836785556286597e-08, \"effect mgl\": -72.46850095526497, \"effect t3\": -31.30639022573031}, {\"Name\": \"Toll Like Receptor 5 (TLR5) Cascade\", \"min fdr\": 3.91149932509524e-08, \"effect mgl\": -71.83086031790526, \"effect t3\": -30.76967986163335}, {\"Name\": \"Toll Like Receptor 10 (TLR10) Cascade\", \"min fdr\": 4.36513461864978e-08, \"effect mgl\": -72.01934912266147, \"effect t3\": -31.16358215783142}, {\"Name\": \"Toll Like Receptor TLR1:TLR2 Cascade\", \"min fdr\": 4.084024882793241e-08, \"effect mgl\": -71.31388304748813, \"effect t3\": -31.407879135575083}, {\"Name\": \"Toll Like Receptor 7/8 (TLR7/8) Cascade\", \"min fdr\": 5.177780973514975e-08, \"effect mgl\": -40.88989880660101, \"effect t3\": -2.435453955391272}, {\"Name\": \"MAPK family signaling cascades\", \"min fdr\": 2.509111539125899e-16, \"effect mgl\": 62.89284241753447, \"effect t3\": 25.65106709405495}, {\"Name\": \"KEAP1-NFE2L2 pathway\", \"min fdr\": 1.6896806491706367e-19, \"effect mgl\": 51.674848125528214, \"effect t3\": 16.024395688675376}, {\"Name\": \"Sensory Perception\", \"min fdr\": 0.03613964247562318, \"effect mgl\": 76.00700000000008, \"effect t3\": 40.66800000000005}, {\"Name\": \"Toll Like Receptor 9 (TLR9) Cascade\", \"min fdr\": 2.451668405278611e-07, \"effect mgl\": -39.19479330471407, \"effect t3\": -3.933903832597863}, {\"Name\": \"MyD88 dependent cascade initiated on endosome\", \"min fdr\": 5.177780973514975e-08, \"effect mgl\": -38.65584135908195, \"effect t3\": -3.7542412512087373}, {\"Name\": \"TRAF6 mediated induction of NFkB and MAP kinases upon TLR7/8 or 9 activation\", \"min fdr\": 4.928985953638884e-08, \"effect mgl\": -39.28074759794334, \"effect t3\": -4.483932114513053}, {\"Name\": \"Regulation of PTEN stability and activity\", \"min fdr\": 4.501264948594763e-24, \"effect mgl\": 52.670104697406295, \"effect t3\": 20.40536743515848}, {\"Name\": \"Signaling by VEGF\", \"min fdr\": 0.00011757760460856093, \"effect mgl\": 19.466544223736697, \"effect t3\": 50.48724002659571}, {\"Name\": \"VEGFA-VEGFR2 Pathway\", \"min fdr\": 0.00019312699120890766, \"effect mgl\": 18.682619152498095, \"effect t3\": 48.54642391717323}, {\"Name\": \"TAK1-dependent IKK and NF-kappa-B activation\", \"min fdr\": 1.0541577576661245e-05, \"effect mgl\": -72.18826666666665, \"effect t3\": -42.52333333333333}, {\"Name\": \"Oxygen-dependent proline hydroxylation of Hypoxia-inducible Factor Alpha\", \"min fdr\": 2.1561402934291693e-23, \"effect mgl\": 25.560475322307404, \"effect t3\": -4.102195667154785}, {\"Name\": \"Regulation of IGF Activity by IGFBP\", \"min fdr\": 2.4881761154671095e-09, \"effect mgl\": 25.8086, \"effect t3\": -3.803749999999999}, {\"Name\": \"Post-translational protein phosphorylation\", \"min fdr\": 4.762770308326503e-10, \"effect mgl\": 25.48559999999999, \"effect t3\": -4.014}, {\"Name\": \"APC/C-mediated degradation of cell cycle proteins\", \"min fdr\": 1.1930212110509213e-18, \"effect mgl\": 59.46909946912703, \"effect t3\": 31.695174158905136}, {\"Name\": \"PTEN Regulation\", \"min fdr\": 3.1498002869098147e-25, \"effect mgl\": 46.83256446540877, \"effect t3\": 19.74532075471696}, {\"Name\": \"Signaling by EGFR\", \"min fdr\": 0.03221216998415962, \"effect mgl\": 14.019548428348775, \"effect t3\": -12.640048572528785}, {\"Name\": \"Protein ubiquitination\", \"min fdr\": 1.036324933183614e-06, \"effect mgl\": -29.10023333333333, \"effect t3\": -2.4493666666666662}, {\"Name\": \"Semaphorin interactions\", \"min fdr\": 0.0070016254102366745, \"effect mgl\": 27.08265584415585, \"effect t3\": 0.4557662337662325}, {\"Name\": \"PI3K/AKT Signaling\", \"min fdr\": 3.2417112682495427e-19, \"effect mgl\": 47.38379192872114, \"effect t3\": 20.878987421383634}, {\"Name\": \"Signaling by ERBB2\", \"min fdr\": 0.0017882762500329942, \"effect mgl\": 23.168012948028682, \"effect t3\": -3.3050337813620114}, {\"Name\": \"Transcriptional regulation by RUNX2\", \"min fdr\": 6.231585778703139e-24, \"effect mgl\": 40.46837476190479, \"effect t3\": 14.58821428571429}, {\"Name\": \"Regulation of mitotic cell cycle\", \"min fdr\": 1.3207238562094312e-18, \"effect mgl\": 57.75511307069453, \"effect t3\": 33.09395426855967}, {\"Name\": \"GTP hydrolysis and joining of the 60S ribosomal subunit\", \"min fdr\": 5.254664282936425e-44, \"effect mgl\": -27.8139, \"effect t3\": -4.5489999999999995}, {\"Name\": \"Cap-dependent Translation Initiation\", \"min fdr\": 2.2882746932162775e-45, \"effect mgl\": -27.9593, \"effect t3\": -4.746833333333333}, {\"Name\": \"Eukaryotic Translation Initiation\", \"min fdr\": 2.2882746932162775e-45, \"effect mgl\": -27.950899999999997, \"effect t3\": -4.851666666666667}, {\"Name\": \"Signaling by FGFR2\", \"min fdr\": 0.00031132743187849696, \"effect mgl\": 18.955821115317182, \"effect t3\": -3.954684869675777}, {\"Name\": \"Degradation of GLI1 by the proteasome\", \"min fdr\": 1.8652976740411883e-22, \"effect mgl\": 37.932287083656796, \"effect t3\": 15.286934539984461}, {\"Name\": \"Regulation of NF-kappa B signaling\", \"min fdr\": 0.0004104984554235268, \"effect mgl\": 26.393749999999997, \"effect t3\": 3.913750000000002}, {\"Name\": \"Selective autophagy\", \"min fdr\": 2.2554936728372544e-09, \"effect mgl\": -19.478113095238104, \"effect t3\": 1.9607142857142854}, {\"Name\": \"Synthesis of DNA\", \"min fdr\": 9.280670566797454e-19, \"effect mgl\": 35.40507969230769, \"effect t3\": 14.82767884615384}, {\"Name\": \"RNA polymerase II transcribes snRNA genes\", \"min fdr\": 2.4220661291451466e-06, \"effect mgl\": -21.299473913043474, \"effect t3\": -1.2148478260869562}, {\"Name\": \"Signaling by FGFR3\", \"min fdr\": 0.04202851489487649, \"effect mgl\": 17.29612413444398, \"effect t3\": -2.7430875582750565}, {\"Name\": \"Switching of origins to a post-replicative state\", \"min fdr\": 1.7304492425667548e-20, \"effect mgl\": 32.92461776923076, \"effect t3\": 13.000682692307684}, {\"Name\": \"Cellular response to hypoxia\", \"min fdr\": 9.22031611220374e-22, \"effect mgl\": 17.729277544167864, \"effect t3\": -1.9891592251661638}, {\"Name\": \"Nonsense-Mediated Decay (NMD)\", \"min fdr\": 3.229220160969386e-35, \"effect mgl\": -26.463644676941936, \"effect t3\": -6.8169506991936295}, {\"Name\": \"Mitophagy\", \"min fdr\": 3.4816853561566065e-08, \"effect mgl\": -19.088204166666667, \"effect t3\": 0.0}, {\"Name\": \"PINK1-PRKN Mediated Mitophagy\", \"min fdr\": 1.1351937002870344e-07, \"effect mgl\": -19.02925, \"effect t3\": 0.0}, {\"Name\": \"Nonsense Mediated Decay (NMD) enhanced by the Exon Junction Complex (EJC)\", \"min fdr\": 3.229220160969386e-35, \"effect mgl\": -26.029500454731053, \"effect t3\": -7.037140042870266}, {\"Name\": \"S Phase\", \"min fdr\": 8.41829036214244e-21, \"effect mgl\": 29.865527018759757, \"effect t3\": 11.080465263870908}, {\"Name\": \"MAPK6/MAPK4 signaling\", \"min fdr\": 1.7375816387617108e-21, \"effect mgl\": 27.515361481481488, \"effect t3\": 9.954355555555553}, {\"Name\": \"Retrograde transport at the Trans-Golgi-Network\", \"min fdr\": 0.0018905821514327743, \"effect mgl\": 20.300500000000003, \"effect t3\": 2.8860000000000006}, {\"Name\": \"RAF/MAP kinase cascade\", \"min fdr\": 8.597251461031422e-15, \"effect mgl\": 25.58057307485614, \"effect t3\": 8.214186150730303}, {\"Name\": \"Translation\", \"min fdr\": 1.462838289893074e-66, \"effect mgl\": -25.02359999999999, \"effect t3\": -7.771333333333333}, {\"Name\": \"ERK1/ERK2 pathway\", \"min fdr\": 1.172274720330281e-14, \"effect mgl\": 25.02418399233711, \"effect t3\": 8.072426975607343}, {\"Name\": \"EGFR downregulation\", \"min fdr\": 0.04202851489487649, \"effect mgl\": 5.971861136837803, \"effect t3\": -10.76811996379468}, {\"Name\": \"Signaling by FGFR\", \"min fdr\": 0.001728865918079503, \"effect mgl\": 12.734294214633932, \"effect t3\": -4.001914639376082}, {\"Name\": \"Collagen formation\", \"min fdr\": 0.03720261269039188, \"effect mgl\": 1.3033538037737868, \"effect t3\": -15.30853705655326}, {\"Name\": \"TNFR2 non-canonical NF-kB pathway\", \"min fdr\": 5.021332405651101e-15, \"effect mgl\": 20.948019657044927, \"effect t3\": 4.4575898664947005}, {\"Name\": \"G1/S Transition\", \"min fdr\": 3.756534309122441e-18, \"effect mgl\": 24.20683197430476, \"effect t3\": 7.908994707141079}, {\"Name\": \"Autodegradation of the E3 ubiquitin ligase COP1\", \"min fdr\": 2.452161665915433e-22, \"effect mgl\": 23.162675555555563, \"effect t3\": 7.011799999999997}, {\"Name\": \"Negative regulation of FGFR3 signaling\", \"min fdr\": 0.019919940343671116, \"effect mgl\": 12.259723958333327, \"effect t3\": -3.821531249999998}, {\"Name\": \"Signaling by FGFR4\", \"min fdr\": 0.03139761488170759, \"effect mgl\": 13.79199192176871, \"effect t3\": -2.164809523809525}, {\"Name\": \"Ubiquitin Mediated Degradation of Phosphorylated Cdc25A\", \"min fdr\": 4.093084066012807e-21, \"effect mgl\": 22.661964444444454, \"effect t3\": 6.881244444444442}, {\"Name\": \"p53-Independent G1/S DNA damage checkpoint\", \"min fdr\": 4.156652502135802e-21, \"effect mgl\": 22.235275555555567, \"effect t3\": 6.731799999999998}, {\"Name\": \"p53-Independent DNA Damage Response\", \"min fdr\": 4.0470222969265035e-21, \"effect mgl\": 22.564194777777786, \"effect t3\": 7.1347111111111055}, {\"Name\": \"Cell Cycle, Mitotic\", \"min fdr\": 6.273038911145665e-26, \"effect mgl\": 68.1932431241581, \"effect t3\": 53.38714325332368}, {\"Name\": \"Transport of Mature Transcript to Cytoplasm\", \"min fdr\": 2.1472034778763692e-10, \"effect mgl\": -11.157763277335716, \"effect t3\": -25.761057618626566}, {\"Name\": \"SCF(Skp2)-mediated degradation of p27/p21\", \"min fdr\": 1.8974344226331904e-19, \"effect mgl\": 20.889603444462512, \"effect t3\": 6.957111074971551}, {\"Name\": \"Negative regulation of FGFR1 signaling\", \"min fdr\": 0.039916604277611505, \"effect mgl\": 10.18737414965986, \"effect t3\": -3.6456700680272123}, {\"Name\": \"Regulation of RAS by GAPs\", \"min fdr\": 2.629296746519711e-18, \"effect mgl\": 21.699913803888414, \"effect t3\": 8.162767857142857}, {\"Name\": \"CDK-mediated phosphorylation and removal of Cdc6\", \"min fdr\": 3.1757214358323237e-21, \"effect mgl\": 21.6479080952381, \"effect t3\": 8.248523809523807}, {\"Name\": \"RNA Polymerase II Transcription Elongation\", \"min fdr\": 5.3204092588088695e-11, \"effect mgl\": -12.386999999999999, \"effect t3\": 0.97375}, {\"Name\": \"Cyclin A:Cdk2-associated events at S phase entry\", \"min fdr\": 4.2461638074730415e-20, \"effect mgl\": 21.23945101915185, \"effect t3\": 8.058274465958121}, {\"Name\": \"Cyclin E associated events during G1/S transition\", \"min fdr\": 7.193826525016272e-20, \"effect mgl\": 20.782534859326713, \"effect t3\": 7.960082114978045}, {\"Name\": \"TCR signaling\", \"min fdr\": 4.393449617294108e-20, \"effect mgl\": 24.328996484281255, \"effect t3\": 11.541192300307218}, {\"Name\": \"Asymmetric localization of PCP proteins\", \"min fdr\": 4.5618464682533025e-17, \"effect mgl\": 20.717775000000003, \"effect t3\": 8.193750000000001}, {\"Name\": \"Regulation of mRNA stability by proteins that bind AU-rich elements\", \"min fdr\": 8.395823660865417e-26, \"effect mgl\": 17.76192, \"effect t3\": 5.492999999999999}, {\"Name\": \"G1/S DNA Damage Checkpoints\", \"min fdr\": 2.2299066160881815e-17, \"effect mgl\": 19.58500757763976, \"effect t3\": 7.406996894409929}, {\"Name\": \"G2/M Checkpoints\", \"min fdr\": 1.900926712352984e-15, \"effect mgl\": 20.07830333333333, \"effect t3\": 7.903306413459934}, {\"Name\": \"Regulated Necrosis\", \"min fdr\": 9.605348884319772e-09, \"effect mgl\": 19.77234210526316, \"effect t3\": 7.603289473684212}, {\"Name\": \"Processing of Capped Intron-Containing Pre-mRNA\", \"min fdr\": 4.2358851566764454e-27, \"effect mgl\": -26.048167724639224, \"effect t3\": -14.099420685811207}, {\"Name\": \"Downstream TCR signaling\", \"min fdr\": 3.354840670064482e-19, \"effect mgl\": 21.597109414575872, \"effect t3\": 9.697841013824885}, {\"Name\": \"SUMOylation of transcription cofactors\", \"min fdr\": 0.0011049225975082084, \"effect mgl\": -12.93521083333333, \"effect t3\": -1.140125}, {\"Name\": \"Neddylation\", \"min fdr\": 1.220296078223956e-26, \"effect mgl\": -10.25676433681577, \"effect t3\": 1.5223490177297299}, {\"Name\": \"AUF1 (hnRNP D0) binds and destabilizes mRNA\", \"min fdr\": 1.3835949447414218e-22, \"effect mgl\": 17.230420000000006, \"effect t3\": 5.8740000000000006}, {\"Name\": \"Formation of Fibrin Clot (Clotting Cascade)\", \"min fdr\": 0.0004480546080243079, \"effect mgl\": 9.278924999999997, \"effect t3\": -2.034583333333333}, {\"Name\": \"Cell Cycle Checkpoints\", \"min fdr\": 8.694284427790727e-18, \"effect mgl\": 20.567272761789933, \"effect t3\": 9.323646844860464}, {\"Name\": \"p53-Dependent G1/S DNA damage checkpoint\", \"min fdr\": 5.354566918007224e-18, \"effect mgl\": 19.069964444444455, \"effect t3\": 7.863938888888887}, {\"Name\": \"Stabilization of p53\", \"min fdr\": 2.6719486753499776e-20, \"effect mgl\": 18.180648571428584, \"effect t3\": 7.075555555555552}, {\"Name\": \"p53-Dependent G1 DNA Damage Response\", \"min fdr\": 5.354566918007224e-18, \"effect mgl\": 18.460984285714297, \"effect t3\": 7.689376190476186}, {\"Name\": \"Cilium Assembly\", \"min fdr\": 4.175368533892241e-05, \"effect mgl\": 11.717173076923073, \"effect t3\": 1.096}, {\"Name\": \"Oxidative Stress Induced Senescence\", \"min fdr\": 0.002191811038357828, \"effect mgl\": -8.127214285714286, \"effect t3\": 2.2585}, {\"Name\": \"MTOR signalling\", \"min fdr\": 2.5939063944583366e-07, \"effect mgl\": -11.572550246305417, \"effect t3\": -1.2956494643834549}, {\"Name\": \"Epigenetic regulation of gene expression\", \"min fdr\": 0.0012769224955931802, \"effect mgl\": -10.216999999999999, \"effect t3\": 0.0}, {\"Name\": \"Amino acid and derivative metabolism\", \"min fdr\": 6.429653644537972e-49, \"effect mgl\": 8.51692666666667, \"effect t3\": -1.6480000000000001}, {\"Name\": \"Organelle biogenesis and maintenance\", \"min fdr\": 7.098129990643775e-09, \"effect mgl\": 9.847442307692301, \"effect t3\": 0.25249999999999995}, {\"Name\": \"Protein localization\", \"min fdr\": 6.189042362576101e-26, \"effect mgl\": -9.925328571428572, \"effect t3\": -0.3400000000000001}, {\"Name\": \"O-glycosylation of TSR domain-containing proteins\", \"min fdr\": 0.03081175007818603, \"effect mgl\": 10.26, \"effect t3\": 0.897}, {\"Name\": \"Peroxisomal protein import\", \"min fdr\": 4.713397361597536e-18, \"effect mgl\": -9.889442857142857, \"effect t3\": -0.546}, {\"Name\": \"Pyroptosis\", \"min fdr\": 0.0008554088543188687, \"effect mgl\": 18.160236842105267, \"effect t3\": 8.915789473684212}, {\"Name\": \"Anchoring of the basal body to the plasma membrane\", \"min fdr\": 0.000475301681674943, \"effect mgl\": 9.646566666666665, \"effect t3\": 0.405}, {\"Name\": \"Regulation of HSF1-mediated heat shock response\", \"min fdr\": 0.001728865918079503, \"effect mgl\": 12.152031761430724, \"effect t3\": 21.34512841958107}, {\"Name\": \"O-linked glycosylation\", \"min fdr\": 0.019151453770010856, \"effect mgl\": 10.329999999999998, \"effect t3\": 1.1675}, {\"Name\": \"FCERI mediated NF-kB activation\", \"min fdr\": 1.3203709383478712e-20, \"effect mgl\": 18.201486501587297, \"effect t3\": 9.048081904761904}, {\"Name\": \"Regulation of ornithine decarboxylase (ODC)\", \"min fdr\": 1.7254880914355715e-20, \"effect mgl\": 12.236310000000003, \"effect t3\": 3.1088333333333336}, {\"Name\": \"RHO GTPases Activate WASPs and WAVEs\", \"min fdr\": 0.001728865918079503, \"effect mgl\": 9.2133, \"effect t3\": 0.09200000000000141}, {\"Name\": \"Negative regulation of FGFR4 signaling\", \"min fdr\": 0.017515521473670664, \"effect mgl\": 5.777538265306121, \"effect t3\": -3.285397959183674}, {\"Name\": \"EPHB-mediated forward signaling\", \"min fdr\": 0.00554964061727294, \"effect mgl\": -8.98299999999999, \"effect t3\": 0.0}, {\"Name\": \"Metabolism of polyamines\", \"min fdr\": 4.393449617294108e-20, \"effect mgl\": 11.48555666666667, \"effect t3\": 2.5890555555555563}, {\"Name\": \"Regulation of actin dynamics for phagocytic cup formation\", \"min fdr\": 0.000663041357833148, \"effect mgl\": 10.902507317073166, \"effect t3\": 2.0566666666666684}, {\"Name\": \"Gap-filling DNA repair synthesis and ligation in TC-NER\", \"min fdr\": 2.3753382556844218e-07, \"effect mgl\": -7.60925, \"effect t3\": 1.2140000000000004}, {\"Name\": \"Death Receptor Signalling\", \"min fdr\": 0.0002963957545667645, \"effect mgl\": -7.664476750663666, \"effect t3\": 0.9409155384867444}, {\"Name\": \"Nucleotide Excision Repair\", \"min fdr\": 4.2610290834180606e-09, \"effect mgl\": -8.617207368421054, \"effect t3\": -0.1453947368421058}, {\"Name\": \"DNA Damage Bypass\", \"min fdr\": 2.3129250439139744e-05, \"effect mgl\": 22.160765797788308, \"effect t3\": 13.748868878357028}, {\"Name\": \"Signaling by ERBB4\", \"min fdr\": 0.0005475838083873893, \"effect mgl\": 12.849374999999998, \"effect t3\": 4.510000000000002}, {\"Name\": \"Ubiquitin-dependent degradation of Cyclin D\", \"min fdr\": 8.780460395474699e-23, \"effect mgl\": 12.025709999999993, \"effect t3\": 3.688711111111112}, {\"Name\": \"Cellular Senescence\", \"min fdr\": 0.0002934463991287297, \"effect mgl\": -7.8700125, \"effect t3\": 0.4649000000000004}, {\"Name\": \"Formation of the Early Elongation Complex\", \"min fdr\": 9.524468576480038e-07, \"effect mgl\": -7.33425, \"effect t3\": 0.779}, {\"Name\": \"The role of GTSE1 in G2/M progression after G2 checkpoint\", \"min fdr\": 4.898004017105356e-22, \"effect mgl\": 11.624768876811595, \"effect t3\": 3.6174393115942025}, {\"Name\": \"SUMOylation of DNA replication proteins\", \"min fdr\": 0.037789300324191005, \"effect mgl\": 10.5410329887218, \"effect t3\": 2.717876879699248}, {\"Name\": \"RHO GTPase Effectors\", \"min fdr\": 2.37187919178843e-12, \"effect mgl\": 12.741221205388293, \"effect t3\": 5.098532703626182}, {\"Name\": \"APC/C:Cdh1 mediated degradation of Cdc20 and other APC/C:Cdh1 targeted proteins in late mitosis/early G1\", \"min fdr\": 1.1415676755247398e-20, \"effect mgl\": 18.246631601519894, \"effect t3\": 10.72517106887349}, {\"Name\": \"Intrinsic Pathway of Fibrin Clot Formation\", \"min fdr\": 0.0014671570840383303, \"effect mgl\": 4.8583720238095225, \"effect t3\": -2.660979166666666}, {\"Name\": \"Cellular response to heat stress\", \"min fdr\": 0.00019915088470509666, \"effect mgl\": 14.123244724547853, \"effect t3\": 21.630867982999032}, {\"Name\": \"Transcription-Coupled Nucleotide Excision Repair (TC-NER)\", \"min fdr\": 1.4762725241586927e-08, \"effect mgl\": -5.708741379310343, \"effect t3\": 1.7616379310344839}, {\"Name\": \"DNA Repair\", \"min fdr\": 6.365966596864935e-08, \"effect mgl\": 3.5257101069250405, \"effect t3\": 10.887477636169145}, {\"Name\": \"Cytokine Signaling in Immune system\", \"min fdr\": 4.828115597185278e-17, \"effect mgl\": -1.7430704806963928, \"effect t3\": 5.488965082294095}, {\"Name\": \"Fcgamma receptor (FCGR) dependent phagocytosis\", \"min fdr\": 0.00269255717235308, \"effect mgl\": 16.608028890824617, \"effect t3\": 9.388195702671311}, {\"Name\": \"Cyclin D associated events in G1\", \"min fdr\": 2.4241285591290627e-05, \"effect mgl\": -6.13602926467472, \"effect t3\": 1.0098879564879566}, {\"Name\": \"G1 Phase\", \"min fdr\": 2.4556812568522125e-05, \"effect mgl\": -5.975320343293072, \"effect t3\": 1.0098879564879566}, {\"Name\": \"Signaling by Hippo\", \"min fdr\": 0.00875835191259607, \"effect mgl\": -6.9609999999999985, \"effect t3\": 0.0}, {\"Name\": \"Centrosome maturation\", \"min fdr\": 0.004948286591757807, \"effect mgl\": 7.719, \"effect t3\": 0.8085}, {\"Name\": \"TCF dependent signaling in response to WNT\", \"min fdr\": 9.125616125365556e-18, \"effect mgl\": 27.10738641425334, \"effect t3\": 21.043440704943585}, {\"Name\": \"G alpha (q) signalling events\", \"min fdr\": 0.0484399088753294, \"effect mgl\": 3.1885000000000003, \"effect t3\": 9.238499999999998}, {\"Name\": \"Cell Cycle\", \"min fdr\": 4.1990691812233364e-30, \"effect mgl\": 52.70986464044206, \"effect t3\": 46.740069904939276}, {\"Name\": \"FCERI mediated Ca+2 mobilization\", \"min fdr\": 0.026736524614117095, \"effect mgl\": 1.568, \"effect t3\": 7.441500000000001}, {\"Name\": \"Clathrin-mediated endocytosis\", \"min fdr\": 2.2242062845442955e-09, \"effect mgl\": 3.908115218538525, \"effect t3\": -1.9105824641460232}, {\"Name\": \"mRNA 3'-end processing\", \"min fdr\": 6.6560424857432e-11, \"effect mgl\": -9.665916666666666, \"effect t3\": -3.8664999999999994}, {\"Name\": \"NIK-->noncanonical NF-kB signaling\", \"min fdr\": 2.4734084199373034e-24, \"effect mgl\": 6.346061818181816, \"effect t3\": 0.5749527272727282}, {\"Name\": \"SUMOylation of RNA binding proteins\", \"min fdr\": 0.00031132743187849696, \"effect mgl\": -7.220306630221474, \"effect t3\": -1.4819918699186987}, {\"Name\": \"Dectin-1 mediated noncanonical NF-kB signaling\", \"min fdr\": 2.25620331960437e-24, \"effect mgl\": 6.172726181818183, \"effect t3\": 0.5749527272727282}, {\"Name\": \"Platelet Aggregation (Plug Formation)\", \"min fdr\": 0.0011273499548642841, \"effect mgl\": -6.551208333333333, \"effect t3\": -0.9583333333333334}, {\"Name\": \"Regulation of innate immune responses to cytosolic DNA\", \"min fdr\": 0.0014892491076335104, \"effect mgl\": 4.202516666666667, \"effect t3\": -1.3786666666666667}, {\"Name\": \"Integrin signaling\", \"min fdr\": 0.0011273499548642841, \"effect mgl\": -6.427833333333334, \"effect t3\": -0.9583333333333334}, {\"Name\": \"Mitotic G1 phase and G1/S transition\", \"min fdr\": 4.4648032357393715e-18, \"effect mgl\": 14.50032455721446, \"effect t3\": 9.056146017429072}, {\"Name\": \"Regulation of TP53 Activity\", \"min fdr\": 7.066447228973573e-10, \"effect mgl\": -3.87824244053302, \"effect t3\": 1.5470200298891614}, {\"Name\": \"RNA Polymerase II Transcription Termination\", \"min fdr\": 6.659224193863673e-12, \"effect mgl\": -8.696450000000002, \"effect t3\": -3.332}, {\"Name\": \"Loss of Nlp from mitotic centrosomes\", \"min fdr\": 0.0024373576402937952, \"effect mgl\": 5.397650000000001, \"effect t3\": 0.2280000000000002}, {\"Name\": \"Signaling by TGF-beta Receptor Complex\", \"min fdr\": 4.549720795920499e-05, \"effect mgl\": 6.46405882967607, \"effect t3\": 1.2962595611285268}, {\"Name\": \"Signaling by TGFB family members\", \"min fdr\": 2.3668275775499944e-05, \"effect mgl\": 6.133715923891997, \"effect t3\": 1.141657936424121}, {\"Name\": \"Recruitment of mitotic centrosome proteins and complexes\", \"min fdr\": 0.00481658736547936, \"effect mgl\": 5.451500000000001, \"effect t3\": 0.5185000000000002}, {\"Name\": \"Loss of proteins required for interphase microtubule organization from the centrosome\", \"min fdr\": 0.002647867856915328, \"effect mgl\": 5.135000000000001, \"effect t3\": 0.20650000000000057}, {\"Name\": \"TGF-beta receptor signaling activates SMADs\", \"min fdr\": 3.1349712267968965e-05, \"effect mgl\": 5.417416666666666, \"effect t3\": 0.549}, {\"Name\": \"SUMOylation of DNA damage response and repair proteins\", \"min fdr\": 0.00034744615959930047, \"effect mgl\": 0.31209202456348395, \"effect t3\": 5.125330474556891}, {\"Name\": \"Signaling by NTRK1 (TRKA)\", \"min fdr\": 0.0006758879309092685, \"effect mgl\": 4.7832, \"effect t3\": 0.0}, {\"Name\": \"Mitochondrial translation\", \"min fdr\": 2.2140725460152158e-17, \"effect mgl\": 2.9469999999999996, \"effect t3\": -1.6945000000000001}, {\"Name\": \"Chromatin modifying enzymes\", \"min fdr\": 0.029515006275480846, \"effect mgl\": 3.2448207271693152, \"effect t3\": -1.359}, {\"Name\": \"Chylomicron assembly\", \"min fdr\": 1.4083773011282292e-07, \"effect mgl\": 10.891, \"effect t3\": 6.328999999999999}, {\"Name\": \"Formation of RNA Pol II elongation complex\", \"min fdr\": 1.3927370529497826e-10, \"effect mgl\": -3.6956499999999997, \"effect t3\": 0.7789999999999999}, {\"Name\": \"Chromatin organization\", \"min fdr\": 0.02958635888754487, \"effect mgl\": 3.1024176801737755, \"effect t3\": -1.359}, {\"Name\": \"Regulation of insulin secretion\", \"min fdr\": 0.018713976383709415, \"effect mgl\": -1.0215, \"effect t3\": 3.3860000000000006}, {\"Name\": \"Complement cascade\", \"min fdr\": 4.420714480060916e-11, \"effect mgl\": -0.39913084093164886, \"effect t3\": 3.98484720575188}, {\"Name\": \"Mitochondrial translation termination\", \"min fdr\": 8.397513970367976e-17, \"effect mgl\": 2.75335, \"effect t3\": -1.6245000000000003}, {\"Name\": \"Fc epsilon receptor (FCERI) signaling\", \"min fdr\": 2.813913221391529e-19, \"effect mgl\": 25.709338243386245, \"effect t3\": 21.40771693121693}, {\"Name\": \"p75NTR recruits signalling complexes\", \"min fdr\": 0.0020572470104315605, \"effect mgl\": -4.261, \"effect t3\": 0.0}, {\"Name\": \"p75NTR signals via NF-kB\", \"min fdr\": 0.02083038759258248, \"effect mgl\": -4.261, \"effect t3\": 0.0}, {\"Name\": \"Integration of energy metabolism\", \"min fdr\": 0.016281070446327237, \"effect mgl\": -1.063, \"effect t3\": 3.1875}, {\"Name\": \"Autodegradation of Cdh1 by Cdh1:APC/C\", \"min fdr\": 2.2722610092577546e-23, \"effect mgl\": 11.146359112187255, \"effect t3\": 6.905245964487487}, {\"Name\": \"Signaling by NTRKs\", \"min fdr\": 0.0005020924637641798, \"effect mgl\": 4.2009, \"effect t3\": 0.0}, {\"Name\": \"Binding and Uptake of Ligands by Scavenger Receptors\", \"min fdr\": 0.010435497946840042, \"effect mgl\": 2.722, \"effect t3\": -1.478}, {\"Name\": \"Chylomicron remodeling\", \"min fdr\": 8.734065255105588e-07, \"effect mgl\": 5.253772727272727, \"effect t3\": 1.103416666666667}, {\"Name\": \"Deadenylation-dependent mRNA decay\", \"min fdr\": 2.9598672548203474e-09, \"effect mgl\": -4.5677, \"effect t3\": -0.581}, {\"Name\": \"Cargo recognition for clathrin-mediated endocytosis\", \"min fdr\": 8.130896436983913e-08, \"effect mgl\": -4.1312, \"effect t3\": -0.31837499999999996}, {\"Name\": \"DNA Damage Recognition in GG-NER\", \"min fdr\": 0.0001539378380569306, \"effect mgl\": 4.621099999999999, \"effect t3\": 0.8780000000000001}, {\"Name\": \"DDX58/IFIH1-mediated induction of interferon-alpha/beta\", \"min fdr\": 1.5562908297816152e-05, \"effect mgl\": 5.2262499999999985, \"effect t3\": 1.5784999999999998}, {\"Name\": \"UCH proteinases\", \"min fdr\": 2.6231676663384765e-15, \"effect mgl\": 3.191166666666667, \"effect t3\": -0.4259999999999999}, {\"Name\": \"Deadenylation of mRNA\", \"min fdr\": 2.7224112237494054e-05, \"effect mgl\": -4.093199999999999, \"effect t3\": -0.581}, {\"Name\": \"VEGFR2 mediated vascular permeability\", \"min fdr\": 0.006788994515483378, \"effect mgl\": -1.0809285714285712, \"effect t3\": 2.39}, {\"Name\": \"Activation of BAD and translocation to mitochondria\", \"min fdr\": 1.9107855926035084e-05, \"effect mgl\": 4.2725, \"effect t3\": 0.8240000000000001}, {\"Name\": \"Smooth Muscle Contraction\", \"min fdr\": 0.008662973336563838, \"effect mgl\": 4.9735, \"effect t3\": 1.5265000000000002}, {\"Name\": \"IRAK2 mediated activation of TAK1 complex upon TLR7/8 or 9 stimulation\", \"min fdr\": 0.0007907766965509038, \"effect mgl\": 0.4009499999999994, \"effect t3\": 3.7895}, {\"Name\": \"Cytosolic sensors of pathogen-associated DNA\", \"min fdr\": 3.231129176897827e-05, \"effect mgl\": 27.622745104895092, \"effect t3\": 24.271749999999987}, {\"Name\": \"Nuclear Envelope (NE) Reassembly\", \"min fdr\": 0.041034384180390925, \"effect mgl\": -2.7233333333333327, \"effect t3\": 0.593}, {\"Name\": \"Regulation of PLK1 Activity at G2/M Transition\", \"min fdr\": 9.066678158297335e-06, \"effect mgl\": 3.8297499999999998, \"effect t3\": 0.68625}, {\"Name\": \"Fanconi Anemia Pathway\", \"min fdr\": 0.027743781796426472, \"effect mgl\": 5.066612876254182, \"effect t3\": 1.9274899665551843}, {\"Name\": \"NRIF signals cell death from the nucleus\", \"min fdr\": 0.042273821506606316, \"effect mgl\": 5.664999999999994, \"effect t3\": 8.784}, {\"Name\": \"Common Pathway of Fibrin Clot Formation\", \"min fdr\": 0.0005527784768508403, \"effect mgl\": 3.623016666666666, \"effect t3\": 0.548}, {\"Name\": \"Regulation of TP53 Activity through Acetylation\", \"min fdr\": 0.0003602172336708672, \"effect mgl\": -2.4295721153846155, \"effect t3\": 0.6192307692307693}, {\"Name\": \"Regulation of PTEN gene transcription\", \"min fdr\": 0.0008913950551161311, \"effect mgl\": 2.9444999999999997, \"effect t3\": 0.0}, {\"Name\": \"DAP12 interactions\", \"min fdr\": 0.04675715753425292, \"effect mgl\": 3.0382727272727275, \"effect t3\": 5.982545454545456}, {\"Name\": \"Glycosaminoglycan metabolism\", \"min fdr\": 0.012997360712666037, \"effect mgl\": -2.469, \"effect t3\": 0.4715}, {\"Name\": \"ISG15 antiviral mechanism\", \"min fdr\": 9.605348884319772e-09, \"effect mgl\": 23.549850000000003, \"effect t3\": 20.697200000000002}, {\"Name\": \"Interleukin-1 signaling\", \"min fdr\": 5.42307758932646e-21, \"effect mgl\": -12.558324520920541, \"effect t3\": -15.407329003781452}, {\"Name\": \"Platelet homeostasis\", \"min fdr\": 0.016801507410547836, \"effect mgl\": -2.847, \"effect t3\": 0.0}, {\"Name\": \"Phosphorylation of the APC/C\", \"min fdr\": 0.03004132817547142, \"effect mgl\": -0.5665000000000001, \"effect t3\": 2.2699999999999996}, {\"Name\": \"Regulation of Complement cascade\", \"min fdr\": 2.8963316393907785e-09, \"effect mgl\": -2.9675000000000002, \"effect t3\": -0.17200000000000015}, {\"Name\": \"Gamma carboxylation, hypusine formation and arylsulfatase activation\", \"min fdr\": 7.30003581442742e-06, \"effect mgl\": 1.133, \"effect t3\": -1.629}, {\"Name\": \"Antiviral mechanism by IFN-stimulated genes\", \"min fdr\": 8.966964117147421e-09, \"effect mgl\": 23.542156250000005, \"effect t3\": 20.81165}, {\"Name\": \"RHO GTPases Activate Formins\", \"min fdr\": 0.0008396228745166634, \"effect mgl\": -0.8115000000000018, \"effect t3\": 1.9135}, {\"Name\": \"Intra-Golgi traffic\", \"min fdr\": 2.6107141850741573e-06, \"effect mgl\": -2.648, \"effect t3\": 0.0}, {\"Name\": \"COPI-mediated anterograde transport\", \"min fdr\": 3.612766995402035e-12, \"effect mgl\": -2.5962025, \"effect t3\": 0.0}, {\"Name\": \"Regulation of TP53 Activity through Methylation\", \"min fdr\": 0.015510512209554295, \"effect mgl\": -2.5346666666666664, \"effect t3\": 0.0}, {\"Name\": \"Negative regulation of MET activity\", \"min fdr\": 0.004555506856066651, \"effect mgl\": 3.3360749999999992, \"effect t3\": 0.8235000000000001}, {\"Name\": \"Telomere Maintenance\", \"min fdr\": 0.004555506856066651, \"effect mgl\": 2.4889107142857143, \"effect t3\": 0.0}, {\"Name\": \"Interleukin-1 family signaling\", \"min fdr\": 2.5628917302744867e-18, \"effect mgl\": -9.61089630796001, \"effect t3\": -12.093618587847118}, {\"Name\": \"Chromosome Maintenance\", \"min fdr\": 0.022341928034193954, \"effect mgl\": 2.452223214285714, \"effect t3\": 0.0}, {\"Name\": \"Iron uptake and transport\", \"min fdr\": 2.0599822638999995e-06, \"effect mgl\": -2.4245, \"effect t3\": 0.0}, {\"Name\": \"Negative regulation of MAPK pathway\", \"min fdr\": 0.029515006275480846, \"effect mgl\": 2.424411038961039, \"effect t3\": 0.0}, {\"Name\": \"Negative regulators of RIG-I/MDA5 signaling\", \"min fdr\": 0.0007953071076193831, \"effect mgl\": 4.2505, \"effect t3\": 1.8535}, {\"Name\": \"Terminal pathway of complement\", \"min fdr\": 0.0048797931673126675, \"effect mgl\": 0.7924999999999996, \"effect t3\": 3.18675}, {\"Name\": \"Extension of Telomeres\", \"min fdr\": 0.008442260840399302, \"effect mgl\": 2.357580357142857, \"effect t3\": 0.0}, {\"Name\": \"Termination of translesion DNA synthesis\", \"min fdr\": 6.492608333329387e-05, \"effect mgl\": 9.572238888888888, \"effect t3\": 7.244444444444445}, {\"Name\": \"Regulation of APC/C activators between G1/S and early anaphase\", \"min fdr\": 0.0006758879309092685, \"effect mgl\": 3.1856000000000004, \"effect t3\": 0.8784000000000001}, {\"Name\": \"Regulation of expression of SLITs and ROBOs\", \"min fdr\": 0.04382372564471704, \"effect mgl\": 3.0685000000000002, \"effect t3\": 0.8235000000000001}, {\"Name\": \"SUMO E3 ligases SUMOylate target proteins\", \"min fdr\": 1.3848666546509935e-10, \"effect mgl\": 7.0477648540570925, \"effect t3\": 4.81312366433154}, {\"Name\": \"SUMOylation\", \"min fdr\": 8.145709157941344e-11, \"effect mgl\": 1.1864784967708921, \"effect t3\": 3.4157957553881566}, {\"Name\": \"TNF signaling\", \"min fdr\": 3.262369345477702e-07, \"effect mgl\": 2.8340368692448648, \"effect t3\": 0.6492383551352305}, {\"Name\": \"IRAK2 mediated activation of TAK1 complex\", \"min fdr\": 0.0003127815620314933, \"effect mgl\": -2.1710000000000003, \"effect t3\": 0.0}, {\"Name\": \"Signaling by ROBO receptors\", \"min fdr\": 0.02385434676337701, \"effect mgl\": 2.9865000000000004, \"effect t3\": 0.8235000000000001}, {\"Name\": \"Formation of Incision Complex in GG-NER\", \"min fdr\": 0.0003772615026553663, \"effect mgl\": -3.5348999999999995, \"effect t3\": -1.4009999999999998}, {\"Name\": \"Regulation of TP53 Activity through Phosphorylation\", \"min fdr\": 6.52894594973669e-06, \"effect mgl\": -2.127367346938775, \"effect t3\": 0.0}, {\"Name\": \"Resolution of Sister Chromatid Cohesion\", \"min fdr\": 0.0022062252678278684, \"effect mgl\": 1.3303125000000011, \"effect t3\": 3.4503000000000004}, {\"Name\": \"RIPK1-mediated regulated necrosis\", \"min fdr\": 1.1676603618318465e-05, \"effect mgl\": 2.0985, \"effect t3\": 0.0}, {\"Name\": \"Regulation of necroptotic cell death\", \"min fdr\": 1.1764395954110836e-05, \"effect mgl\": 2.0854999999999997, \"effect t3\": 0.0}, {\"Name\": \"Regulation of TP53 Degradation\", \"min fdr\": 0.015770500962559185, \"effect mgl\": -0.9033191964285714, \"effect t3\": 1.147017857142857}, {\"Name\": \"Interleukin-17 signaling\", \"min fdr\": 6.785223492367292e-06, \"effect mgl\": 4.540986249999996, \"effect t3\": 6.552625}, {\"Name\": \"Senescence-Associated Secretory Phenotype (SASP)\", \"min fdr\": 0.005314146608572757, \"effect mgl\": 1.2168, \"effect t3\": -0.7649499999999998}, {\"Name\": \"Deactivation of the beta-catenin transactivating complex\", \"min fdr\": 0.006612063707999712, \"effect mgl\": -1.9779999999999998, \"effect t3\": 0.0}, {\"Name\": \"Regulation of TP53 Expression and Degradation\", \"min fdr\": 0.016603523741830356, \"effect mgl\": -0.8045825892857141, \"effect t3\": 1.147017857142857}, {\"Name\": \"RNA Polymerase I Transcription\", \"min fdr\": 9.077963411486252e-08, \"effect mgl\": -1.91, \"effect t3\": 0.0}, {\"Name\": \"RNA Polymerase I Transcription Initiation\", \"min fdr\": 3.1371853253442665e-07, \"effect mgl\": -1.91, \"effect t3\": 0.0}, {\"Name\": \"RNA Polymerase I Promoter Clearance\", \"min fdr\": 3.146326157722632e-07, \"effect mgl\": -1.91, \"effect t3\": 0.0}, {\"Name\": \"Regulation of TNFR1 signaling\", \"min fdr\": 4.74838957663425e-06, \"effect mgl\": 2.6438048552203, \"effect t3\": 0.8377658649200043}, {\"Name\": \"HS-GAG degradation\", \"min fdr\": 0.04626376542368423, \"effect mgl\": -0.9385000000000002, \"effect t3\": 0.8350000000000001}, {\"Name\": \"Translesion synthesis by Y family DNA polymerases bypasses lesions on DNA template\", \"min fdr\": 1.2151032344734676e-05, \"effect mgl\": 8.281055555555554, \"effect t3\": 6.569925925925925}, {\"Name\": \"Inactivation of CSF3 (G-CSF) signaling\", \"min fdr\": 0.004555506856066651, \"effect mgl\": -1.708076923076923, \"effect t3\": 0.0}, {\"Name\": \"Downregulation of ERBB4 signaling\", \"min fdr\": 0.0005520862013613292, \"effect mgl\": 5.494999999999998, \"effect t3\": 3.842999999999999}, {\"Name\": \"HS-GAG biosynthesis\", \"min fdr\": 0.04708994453699301, \"effect mgl\": -0.8135000000000001, \"effect t3\": 0.835}, {\"Name\": \"Signaling by CSF3 (G-CSF)\", \"min fdr\": 0.004992822534046836, \"effect mgl\": -1.6410000000000002, \"effect t3\": 0.0}, {\"Name\": \"ALPK1 signaling pathway\", \"min fdr\": 0.000538555586430029, \"effect mgl\": -1.6149999999999998, \"effect t3\": 0.0}, {\"Name\": \"G2/M DNA damage checkpoint\", \"min fdr\": 0.01532593130922764, \"effect mgl\": -1.613793103448276, \"effect t3\": 0.0}, {\"Name\": \"TRAF6-mediated induction of TAK1 complex within TLR4 complex\", \"min fdr\": 0.0012734946152540335, \"effect mgl\": 0.469111111111111, \"effect t3\": 2.025}, {\"Name\": \"MAP kinase activation\", \"min fdr\": 6.785223492367292e-06, \"effect mgl\": 5.011567499999996, \"effect t3\": 6.552625}, {\"Name\": \"Transcriptional regulation by RUNX3\", \"min fdr\": 4.2461638074730415e-20, \"effect mgl\": 2.089799999999999, \"effect t3\": 0.5489999999999998}, {\"Name\": \"Pexophagy\", \"min fdr\": 0.041352618281719225, \"effect mgl\": 3.492607142857141, \"effect t3\": 1.9607142857142854}, {\"Name\": \"Vitamin B5 (pantothenate) metabolism\", \"min fdr\": 0.018800601835807992, \"effect mgl\": 1.53, \"effect t3\": 0.0}, {\"Name\": \"SCF-beta-TrCP mediated degradation of Emi1\", \"min fdr\": 0.0025899955771532777, \"effect mgl\": 2.0408, \"effect t3\": 0.549}, {\"Name\": \"Mitochondrial biogenesis\", \"min fdr\": 7.460864571153127e-07, \"effect mgl\": -1.9916, \"effect t3\": -0.5310000000000001}, {\"Name\": \"Regulation of RUNX3 expression and activity\", \"min fdr\": 1.897146593306539e-20, \"effect mgl\": 1.992, \"effect t3\": 0.5489999999999998}, {\"Name\": \"Synthesis of active ubiquitin: roles of E1 and E2 enzymes\", \"min fdr\": 8.166597364678615e-08, \"effect mgl\": 1.991, \"effect t3\": 0.549}, {\"Name\": \"Formation of TC-NER Pre-Incision Complex\", \"min fdr\": 2.3233522423439938e-08, \"effect mgl\": 1.9909999999999997, \"effect t3\": 0.549}, {\"Name\": \"Recognition of DNA damage by PCNA-containing replication complex\", \"min fdr\": 0.007065853410325358, \"effect mgl\": 1.9909999999999999, \"effect t3\": 0.549}, {\"Name\": \"Signaling by NOTCH1\", \"min fdr\": 0.00813882714990321, \"effect mgl\": 1.9909999999999999, \"effect t3\": 0.549}, {\"Name\": \"Activated NOTCH1 Transmits Signal to the Nucleus\", \"min fdr\": 0.024467379858921164, \"effect mgl\": 1.9909999999999999, \"effect t3\": 0.549}, {\"Name\": \"FLT3 Signaling\", \"min fdr\": 0.03081175007818603, \"effect mgl\": 1.4412091619318177, \"effect t3\": 0.0}, {\"Name\": \"Downregulation of TGF-beta receptor signaling\", \"min fdr\": 0.002189217364950244, \"effect mgl\": 1.9895, \"effect t3\": 0.549}, {\"Name\": \"HDL remodeling\", \"min fdr\": 0.00026231500009762, \"effect mgl\": 1.4399999999999995, \"effect t3\": 0.0}, {\"Name\": \"Phospholipid metabolism\", \"min fdr\": 1.2960961780953805e-05, \"effect mgl\": 1.04, \"effect t3\": -0.39699999999999996}, {\"Name\": \"Signal amplification\", \"min fdr\": 0.025745712918325024, \"effect mgl\": -1.4234999999999998, \"effect t3\": 0.0}, {\"Name\": \"Metabolism of lipids\", \"min fdr\": 4.9336042859290006e-48, \"effect mgl\": 1.0067500000000005, \"effect t3\": -0.41450000000000004}, {\"Name\": \"Transcriptional activation of mitochondrial biogenesis\", \"min fdr\": 0.005547693818399568, \"effect mgl\": -1.943, \"effect t3\": -0.5310000000000001}, {\"Name\": \"Glycogen metabolism\", \"min fdr\": 0.008512205898265108, \"effect mgl\": 1.3969999999999998, \"effect t3\": 0.0}, {\"Name\": \"RHOG GTPase cycle\", \"min fdr\": 0.00016465150070143475, \"effect mgl\": 1.39, \"effect t3\": 0.0}, {\"Name\": \"Interferon Signaling\", \"min fdr\": 2.1178645078562853e-06, \"effect mgl\": 23.592601388888887, \"effect t3\": 22.217966666666666}, {\"Name\": \"Signaling by Interleukins\", \"min fdr\": 8.079400780870695e-14, \"effect mgl\": -11.935011069617476, \"effect t3\": -10.567306000605852}, {\"Name\": \"RHO GTPases activate PKNs\", \"min fdr\": 0.00040391469098356535, \"effect mgl\": 1.3859999999999997, \"effect t3\": 2.751}, {\"Name\": \"Gamma-carboxylation of protein precursors\", \"min fdr\": 0.0005602977164136174, \"effect mgl\": 0.7090000000000002, \"effect t3\": -0.652}, {\"Name\": \"Josephin domain DUBs\", \"min fdr\": 5.985585261482658e-07, \"effect mgl\": 1.856, \"effect t3\": 0.5489999999999999}, {\"Name\": \"Inactivation, recovery and regulation of the phototransduction cascade\", \"min fdr\": 0.026292100154208915, \"effect mgl\": -0.9323499999999996, \"effect t3\": -2.2319999999999998}, {\"Name\": \"Eukaryotic Translation Termination\", \"min fdr\": 0.015842251534713032, \"effect mgl\": 0.7019999999999997, \"effect t3\": -0.591}, {\"Name\": \"Removal of aminoterminal propeptides from gamma-carboxylated proteins\", \"min fdr\": 0.0013536788076710293, \"effect mgl\": 0.6059999999999999, \"effect t3\": -0.652}, {\"Name\": \"Negative regulation of FLT3\", \"min fdr\": 0.014527280125733774, \"effect mgl\": -1.2471034482758625, \"effect t3\": 0.0}, {\"Name\": \"Transcriptional activity of SMAD2/SMAD3:SMAD4 heterotrimer\", \"min fdr\": 0.0019161715503908312, \"effect mgl\": 2.2642988958990524, \"effect t3\": 1.0330552050473187}, {\"Name\": \"Gamma-carboxylation, transport, and amino-terminal cleavage of proteins\", \"min fdr\": 0.0003381446135966187, \"effect mgl\": 0.5719999999999992, \"effect t3\": -0.652}, {\"Name\": \"Biotin transport and metabolism\", \"min fdr\": 0.032899557136536625, \"effect mgl\": -1.69425, \"effect t3\": -0.4740000000000002}, {\"Name\": \"RHO GTPases activate PAKs\", \"min fdr\": 0.004681830875416604, \"effect mgl\": 1.992333333333333, \"effect t3\": 0.7726666666666666}, {\"Name\": \"Downstream signal transduction\", \"min fdr\": 0.045100896064190855, \"effect mgl\": 1.1599999999999997, \"effect t3\": 0.0}, {\"Name\": \"Regulation of localization of FOXO transcription factors\", \"min fdr\": 0.0005898766146785548, \"effect mgl\": 1.1165000000000003, \"effect t3\": 0.0}, {\"Name\": \"HSF1 activation\", \"min fdr\": 0.015071271450633902, \"effect mgl\": 1.11, \"effect t3\": 0.0}, {\"Name\": \"VLDLR internalisation and degradation\", \"min fdr\": 0.0005951259007157069, \"effect mgl\": -1.0899999999999999, \"effect t3\": 0.0}, {\"Name\": \"Cellular response to starvation\", \"min fdr\": 1.2432174866395055e-05, \"effect mgl\": -0.581749641577061, \"effect t3\": 0.5027198327359617}, {\"Name\": \"FOXO-mediated transcription\", \"min fdr\": 7.705291549087487e-05, \"effect mgl\": 1.0815, \"effect t3\": 0.0}, {\"Name\": \"Amino acids regulate mTORC1\", \"min fdr\": 1.2432174866395055e-05, \"effect mgl\": -0.5926397849462358, \"effect t3\": 0.46234169653524493}, {\"Name\": \"RAB geranylgeranylation\", \"min fdr\": 0.0012734946152540335, \"effect mgl\": 1.05, \"effect t3\": 0.0}, {\"Name\": \"DARPP-32 events\", \"min fdr\": 0.004503701543311814, \"effect mgl\": -2.186, \"effect t3\": -1.1409999999999996}, {\"Name\": \"Signalling to ERKs\", \"min fdr\": 0.007950324845738705, \"effect mgl\": -1.038, \"effect t3\": 0.0}, {\"Name\": \"Downregulation of SMAD2/3:SMAD4 transcriptional activity\", \"min fdr\": 0.004006958337313752, \"effect mgl\": 2.683333333333333, \"effect t3\": 1.647}, {\"Name\": \"Interleukin-3, Interleukin-5 and GM-CSF signaling\", \"min fdr\": 0.028986023276354935, \"effect mgl\": 0.0, \"effect t3\": 1.02}, {\"Name\": \"Glyoxylate metabolism and glycine degradation\", \"min fdr\": 2.6856544709199633e-05, \"effect mgl\": -2.3869999999999996, \"effect t3\": -1.3985}, {\"Name\": \"Heparan sulfate/heparin (HS-GAG) metabolism\", \"min fdr\": 0.033750685644306395, \"effect mgl\": -0.12249999999999994, \"effect t3\": 0.8350000000000002}, {\"Name\": \"SUMOylation of intracellular receptors\", \"min fdr\": 0.03004132817547142, \"effect mgl\": -1.9850020833333324, \"effect t3\": -1.075}, {\"Name\": \"APC/C:Cdc20 mediated degradation of Cyclin B\", \"min fdr\": 0.00033306156585950617, \"effect mgl\": 3.507475000000001, \"effect t3\": 2.6141749999999995}, {\"Name\": \"Glycogen synthesis\", \"min fdr\": 0.026292100154208915, \"effect mgl\": 0.836, \"effect t3\": 0.0}, {\"Name\": \"RHO GTPases activate IQGAPs\", \"min fdr\": 0.00019900589127316737, \"effect mgl\": -0.16943708609271457, \"effect t3\": -0.9971920529801326}, {\"Name\": \"IKK complex recruitment mediated by RIP1\", \"min fdr\": 8.166224356707088e-05, \"effect mgl\": 2.19868076923077, \"effect t3\": 3.0219230769230765}, {\"Name\": \"APC-Cdc20 mediated degradation of Nek2A\", \"min fdr\": 0.0007146546992186053, \"effect mgl\": 3.4916000000000005, \"effect t3\": 2.6725454545454554}, {\"Name\": \"Branched-chain amino acid catabolism\", \"min fdr\": 0.0010386214643882394, \"effect mgl\": -0.8130000000000001, \"effect t3\": 0.0}, {\"Name\": \"Deubiquitination\", \"min fdr\": 7.704603541742005e-28, \"effect mgl\": 1.1932285714285704, \"effect t3\": 0.38322857142857103}, {\"Name\": \"Prolonged ERK activation events\", \"min fdr\": 0.022278428452255774, \"effect mgl\": -0.782, \"effect t3\": 0.0}, {\"Name\": \"Regulation of BACH1 activity\", \"min fdr\": 0.015842251534713032, \"effect mgl\": -0.7746000000000003, \"effect t3\": 0.0}, {\"Name\": \"Glycerophospholipid biosynthesis\", \"min fdr\": 0.0005130896440476576, \"effect mgl\": 0.377, \"effect t3\": -0.39699999999999996}, {\"Name\": \"MAPK targets/ Nuclear events mediated by MAP kinases\", \"min fdr\": 0.01805325909229981, \"effect mgl\": 5.053121428571429, \"effect t3\": 4.288}, {\"Name\": \"HDL assembly\", \"min fdr\": 0.032899557136536625, \"effect mgl\": -0.7220000000000001, \"effect t3\": 0.0}, {\"Name\": \"Respiratory electron transport, ATP synthesis by chemiosmotic coupling, and heat production by uncoupling proteins.\", \"min fdr\": 2.8758676058747243e-30, \"effect mgl\": -0.7128181818181817, \"effect t3\": 0.0}, {\"Name\": \"Respiratory electron transport\", \"min fdr\": 1.0855559098693699e-27, \"effect mgl\": -0.7128181818181817, \"effect t3\": 0.0}, {\"Name\": \"RAC1 GTPase cycle\", \"min fdr\": 0.010742128420708064, \"effect mgl\": -0.6910000000000002, \"effect t3\": 0.0}, {\"Name\": \"Drug ADME\", \"min fdr\": 3.354879025160683e-10, \"effect mgl\": -0.691, \"effect t3\": 0.0}, {\"Name\": \"Azathioprine ADME\", \"min fdr\": 0.0016282009743046964, \"effect mgl\": -0.691, \"effect t3\": 0.0}, {\"Name\": \"Oncogene Induced Senescence\", \"min fdr\": 0.01190476148198459, \"effect mgl\": -0.6898000000000002, \"effect t3\": 0.0}, {\"Name\": \"PI Metabolism\", \"min fdr\": 0.03081175007818603, \"effect mgl\": 0.663, \"effect t3\": 0.0}, {\"Name\": \"Classical antibody-mediated complement activation\", \"min fdr\": 0.0040224051975673285, \"effect mgl\": 2.2637500000000004, \"effect t3\": 1.6090000000000004}, {\"Name\": \"The citric acid (TCA) cycle and respiratory electron transport\", \"min fdr\": 1.7981498821588932e-41, \"effect mgl\": -1.287954545454546, \"effect t3\": -0.6344}, {\"Name\": \"Factors involved in megakaryocyte development and platelet production\", \"min fdr\": 0.04089617263595986, \"effect mgl\": -0.6509999999999999, \"effect t3\": 0.0}, {\"Name\": \"Translesion Synthesis by POLH\", \"min fdr\": 0.0006758879309092685, \"effect mgl\": 0.6439500000000001, \"effect t3\": 0.0}, {\"Name\": \"Mitotic Prometaphase\", \"min fdr\": 2.4326909406553193e-05, \"effect mgl\": 1.9185999999999996, \"effect t3\": 2.5523333333333333}, {\"Name\": \"Global Genome Nucleotide Excision Repair (GG-NER)\", \"min fdr\": 1.5091457387776737e-05, \"effect mgl\": -1.818438461538461, \"effect t3\": -1.1969230769230768}, {\"Name\": \"Signaling by NOTCH3\", \"min fdr\": 0.01805325909229981, \"effect mgl\": -7.611323529411765, \"effect t3\": -7.033558823529413}, {\"Name\": \"SMAD2/SMAD3:SMAD4 heterotrimer regulates transcription\", \"min fdr\": 0.020082761201961466, \"effect mgl\": -0.5550000000000003, \"effect t3\": 0.0}, {\"Name\": \"COPI-dependent Golgi-to-ER retrograde traffic\", \"min fdr\": 9.430058465046761e-06, \"effect mgl\": 0.5389, \"effect t3\": 0.0}, {\"Name\": \"Creation of C4 and C2 activators\", \"min fdr\": 0.00047826704490009953, \"effect mgl\": 2.344825, \"effect t3\": 1.8487500000000003}, {\"Name\": \"Detoxification of Reactive Oxygen Species\", \"min fdr\": 2.390906507873708e-06, \"effect mgl\": -0.4869999999999998, \"effect t3\": 0.0}, {\"Name\": \"Regulation of pyruvate dehydrogenase (PDH) complex\", \"min fdr\": 0.00425373356735154, \"effect mgl\": -0.46880000000000016, \"effect t3\": -0.9452}, {\"Name\": \"Plasma lipoprotein assembly\", \"min fdr\": 3.5991957041568754e-09, \"effect mgl\": 5.874610909090909, \"effect t3\": 5.407636363636364}, {\"Name\": \"RAC2 GTPase cycle\", \"min fdr\": 0.0005529116874912518, \"effect mgl\": 1.6499999999999995, \"effect t3\": 2.1099999999999994}, {\"Name\": \"Formation of the cornified envelope\", \"min fdr\": 0.03813885527193347, \"effect mgl\": 2.17, \"effect t3\": 2.61}, {\"Name\": \"N-glycan trimming in the ER and Calnexin/Calreticulin cycle\", \"min fdr\": 6.268908727086542e-07, \"effect mgl\": 0.4356, \"effect t3\": 0.0}, {\"Name\": \"FCGR activation\", \"min fdr\": 0.009452567303789761, \"effect mgl\": 5.01, \"effect t3\": 5.43}, {\"Name\": \"Nucleotide metabolism\", \"min fdr\": 5.390756513280335e-06, \"effect mgl\": -0.387, \"effect t3\": 0.0}, {\"Name\": \"Interconversion of nucleotide di- and triphosphates\", \"min fdr\": 0.024467379858921164, \"effect mgl\": -0.387, \"effect t3\": 0.0}, {\"Name\": \"SUMOylation of ubiquitinylation proteins\", \"min fdr\": 0.00875835191259607, \"effect mgl\": -1.4429166666666666, \"effect t3\": -1.0610000000000002}, {\"Name\": \"Metabolism of steroids\", \"min fdr\": 2.7235887820197384e-19, \"effect mgl\": -0.36124999999999985, \"effect t3\": 0.0}, {\"Name\": \"Ub-specific processing proteases\", \"min fdr\": 3.8293130074346745e-24, \"effect mgl\": 2.006000000000001, \"effect t3\": 1.6448800000000001}, {\"Name\": \"Ovarian tumor domain proteases\", \"min fdr\": 8.734065255105588e-07, \"effect mgl\": 0.20799999999999974, \"effect t3\": 0.549}, {\"Name\": \"MAP3K8 (TPL2)-dependent MAPK1/3 activation\", \"min fdr\": 7.572480490608987e-05, \"effect mgl\": 2.6044374999999973, \"effect t3\": 2.264625}, {\"Name\": \"MAP2K and MAPK activation\", \"min fdr\": 0.009349001575517218, \"effect mgl\": 0.32450000000000073, \"effect t3\": -0.0050000000000004485}, {\"Name\": \"Antigen Presentation: Folding, assembly and peptide loading of class I MHC\", \"min fdr\": 0.0004976429083394626, \"effect mgl\": 2.2645000000000004, \"effect t3\": 2.0060000000000002}, {\"Name\": \"Initial triggering of complement\", \"min fdr\": 0.00014716812178932058, \"effect mgl\": 2.2523365384615386, \"effect t3\": 2.01225}, {\"Name\": \"Signaling by PTK6\", \"min fdr\": 0.001640616790208001, \"effect mgl\": -3.1606666666666663, \"effect t3\": -3.3975}, {\"Name\": \"Signaling by NOTCH\", \"min fdr\": 0.007422927580865164, \"effect mgl\": -7.091302681992337, \"effect t3\": -7.322645593869733}, {\"Name\": \"NOTCH3 Activation and Transmission of Signal to the Nucleus\", \"min fdr\": 0.01854966933774674, \"effect mgl\": -7.534558823529412, \"effect t3\": -7.327705882352942}, {\"Name\": \"Sulfur amino acid metabolism\", \"min fdr\": 1.0055006732599888e-07, \"effect mgl\": -0.836, \"effect t3\": -0.6599999999999999}, {\"Name\": \"Degradation of cysteine and homocysteine\", \"min fdr\": 1.0238030621389436e-05, \"effect mgl\": -0.836, \"effect t3\": -0.6599999999999999}, {\"Name\": \"mTORC1-mediated signalling\", \"min fdr\": 3.9020828361282084e-05, \"effect mgl\": -0.7635689655172413, \"effect t3\": -0.6010344827586206}, {\"Name\": \"Degradation of beta-catenin by the destruction complex\", \"min fdr\": 1.6302065491356596e-21, \"effect mgl\": -0.531310412186381, \"effect t3\": -0.678579704301075}, {\"Name\": \"Pyruvate metabolism and Citric Acid (TCA) cycle\", \"min fdr\": 7.299922005902439e-11, \"effect mgl\": -0.4944000000000002, \"effect t3\": -0.6344}, {\"Name\": \"Pyruvate metabolism\", \"min fdr\": 0.0008751665259245636, \"effect mgl\": -0.4944000000000002, \"effect t3\": -0.6344}, {\"Name\": \"Signaling by Non-Receptor Tyrosine Kinases\", \"min fdr\": 0.0015774814012380436, \"effect mgl\": -3.340666666666666, \"effect t3\": -3.2015}, {\"Name\": \"Regulation of PTEN localization\", \"min fdr\": 0.0006207671798893156, \"effect mgl\": -0.35372222222222244, \"effect t3\": -0.2195555555555555}, {\"Name\": \"Signaling by Retinoic Acid\", \"min fdr\": 0.00010952340859780226, \"effect mgl\": -0.44142857142857145, \"effect t3\": -0.572}, {\"Name\": \"RHOC GTPase cycle\", \"min fdr\": 0.0003031230022268962, \"effect mgl\": 2.09, \"effect t3\": 2.0}, {\"Name\": \"Activation of the mRNA upon binding of the cap-binding complex and eIFs, and subsequent binding to 43S\", \"min fdr\": 2.871243516951513e-24, \"effect mgl\": -0.6412, \"effect t3\": -0.581}, {\"Name\": \"Protein methylation\", \"min fdr\": 0.00016987415288148246, \"effect mgl\": -1.2031999999999996, \"effect t3\": -1.263}, {\"Name\": \"KSRP (KHSRP) binds and destabilizes mRNA\", \"min fdr\": 0.0021060486205716888, \"effect mgl\": -0.74, \"effect t3\": -0.712}, {\"Name\": \"Glycosphingolipid metabolism\", \"min fdr\": 0.0061572844248615225, \"effect mgl\": 0.5410000000000004, \"effect t3\": 0.531}, {\"Name\": \"Sphingolipid metabolism\", \"min fdr\": 0.005738115105287813, \"effect mgl\": 0.5410000000000004, \"effect t3\": 0.531}, {\"Name\": \"Ion channel transport\", \"min fdr\": 0.025864365154924886, \"effect mgl\": -0.006000000000000227, \"effect t3\": 0.0}, {\"Name\": \"Fatty acid metabolism\", \"min fdr\": 2.4186926762660642e-32, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Biological oxidations\", \"min fdr\": 2.1561402934291693e-23, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Mitochondrial Fatty Acid Beta-Oxidation\", \"min fdr\": 4.0245872673347896e-13, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Phase II - Conjugation of compounds\", \"min fdr\": 1.2431069367921533e-10, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"RHOBTB GTPase Cycle\", \"min fdr\": 4.545749888236555e-07, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"RHOBTB2 GTPase cycle\", \"min fdr\": 6.997495110826782e-07, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"RHOA GTPase cycle\", \"min fdr\": 4.145863419618645e-05, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Aggrephagy\", \"min fdr\": 0.00011400514721785396, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Mitochondrial protein import\", \"min fdr\": 0.0006526251739523326, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Chk1/Chk2(Cds1) mediated inactivation of Cyclin B:Cdk1 complex\", \"min fdr\": 0.0008554088543188687, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Arachidonic acid metabolism\", \"min fdr\": 0.00015283800679605388, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Cooperation of PDCL (PhLP1) and TRiC/CCT in G-protein beta folding\", \"min fdr\": 0.0025120998288759213, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Chaperonin-mediated protein folding\", \"min fdr\": 0.0033683104943663367, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Protein folding\", \"min fdr\": 0.0033683104943663367, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"RAC3 GTPase cycle\", \"min fdr\": 0.009218147169813685, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"RA biosynthesis pathway\", \"min fdr\": 0.009290770131458658, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Receptor Mediated Mitophagy\", \"min fdr\": 0.019919940343671116, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Ion homeostasis\", \"min fdr\": 0.02160528734927951, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Estrogen-dependent gene expression\", \"min fdr\": 0.03081175007818603, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"NF-kB is activated and signals survival\", \"min fdr\": 0.03081175007818603, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Post-translational modification: synthesis of GPI-anchored proteins\", \"min fdr\": 0.031150089570589082, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Conversion from APC/C:Cdc20 to APC/C:Cdh1 in late anaphase\", \"min fdr\": 0.032899557136536625, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Synthesis of Leukotrienes (LT) and Eoxins (EX)\", \"min fdr\": 0.032899557136536625, \"effect mgl\": 0.0, \"effect t3\": 0.0}, {\"Name\": \"Trafficking and processing of endosomal TLR\", \"min fdr\": 0.02497231186015822, \"effect mgl\": 0.0, \"effect t3\": 0.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.LayerChart(...)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/projects/koesterlab/moeller-th-liver-diffexp/analysis_2/.snakemake/conda/673ca4e70a0a7f8fe6fba3391c9fbb49_/lib/python3.12/site-packages/altair/utils/save.py:42: UserWarning: inline argument ignored for non HTML formats.\n",
      "  warnings.warn(\"inline argument ignored for non HTML formats.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "import altair as alt\n",
    "import sys\n",
    "from IPython.display import display\n",
    "\n",
    "# we cannot use vegafusion here because it makes the point selection impossible since\n",
    "# it prunes the required ext_gene column\n",
    "#alt.data_transformers.enable(\"vegafusion\")\n",
    "alt.data_transformers.disable_max_rows()\n",
    "\n",
    "\n",
    "# Data transformation\n",
    "alt.data_transformers.disable_max_rows()\n",
    "\n",
    "# Selektor für Punkte\n",
    "point_selector = alt.selection_point(fields=[\"Name\"], empty=False)\n",
    "\n",
    "# Punkte\n",
    "points = alt.Chart(combined_pd).mark_circle(size=15, tooltip={\"content\": \"data\"}).encode(\n",
    "    alt.X(effect_x, scale=alt.Scale(type='symlog', nice=False), axis=alt.Axis(grid=False)),\n",
    "    alt.Y(effect_y, scale=alt.Scale(type='symlog', nice=False), axis=alt.Axis(grid=False)),\n",
    "    alt.Color(\"min fdr\", scale=alt.Scale(scheme=\"viridis\")),\n",
    "    opacity=alt.value(0.5),\n",
    ")\n",
    "\n",
    "# Gestrichelte Linie\n",
    "line = alt.Chart(\n",
    "    pl.DataFrame({effect_x: [min_value, max_value], effect_y: [min_value, max_value]})\n",
    ").mark_line(color=\"lightgrey\").encode(\n",
    "    x=effect_x,\n",
    "    y=effect_y,\n",
    "    strokeDash=alt.value([5, 5]),\n",
    ")\n",
    "\n",
    "# Gestrichelte Linie\n",
    "x_axis = alt.Chart(\n",
    "    pl.DataFrame({effect_x: [0, 0], effect_y: [min_value, max_value]})\n",
    ").mark_line(color=\"lightgrey\").encode(\n",
    "    x=effect_x,\n",
    "    y=effect_y,\n",
    "    strokeDash=alt.value([5, 5]),\n",
    ")\n",
    "\n",
    "y_axis = alt.Chart(\n",
    "    pl.DataFrame({effect_x: [min_value, max_value], effect_y: [0, 0]})\n",
    ").mark_line(color=\"lightgrey\").encode(\n",
    "    x=effect_x,\n",
    "    y=effect_y,\n",
    "    strokeDash=alt.value([5, 5]),\n",
    ")\n",
    "\n",
    "text_background = alt.Chart(combined_pd).mark_text(\n",
    "        align=\"left\",\n",
    "        baseline=\"middle\",\n",
    "        dx=5,\n",
    "        dy=-5,\n",
    "        fill='white',\n",
    "        stroke='white',\n",
    "        strokeWidth=5,\n",
    "    ).encode(\n",
    "        x=effect_x,\n",
    "        y=effect_y,\n",
    "        text=alt.condition(point_selector, \"Name\", alt.value(\"\")),\n",
    "    )\n",
    "\n",
    "\n",
    "# Text\n",
    "text = alt.Chart(combined_pd).mark_text(\n",
    "    align=\"left\",\n",
    "    baseline=\"middle\",\n",
    "    dx=5,\n",
    "    dy=-5,\n",
    ").encode(\n",
    "    x=effect_x,\n",
    "    y=effect_y,\n",
    "    text=alt.condition(point_selector, \"Name\", alt.value(\"\")),\n",
    ")\n",
    "\n",
    "# Null-Linien für x und y\n",
    "zero_lines = alt.Chart(pl.DataFrame({\"zero\": [0]})).mark_rule(color=\"black\").encode(\n",
    "    x=alt.X(\"zero\", axis=alt.Axis(title=\"\")),  # keine Beschriftung der Achse\n",
    "    y=alt.Y(\"zero\", axis=alt.Axis(title=\"\")),  # keine Beschriftung der Achse\n",
    ")\n",
    "\n",
    "# Gesamtes Diagramm mit den Ebenen\n",
    "chart = alt.layer(x_axis, y_axis, line, points, text_background, text).add_params(\n",
    "    point_selector\n",
    ").interactive()\n",
    "\n",
    "# display(chart)\n",
    "\n",
    "chart.save(snakemake.output[1], inline=True)\n",
    "# chart.save(\"/projects/koesterlab/moeller-th-liver-diffexp/analysis_2/results/meta_comparison/pathways/etoh_wt_mgl_vs_t3.json\", inline=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
